# Comparing `tmp/biocolabsdk-1.0.1.tar.gz` & `tmp/biocolabsdk-1.0.2.tar.gz`

## filetype from file(1)

```diff
@@ -1 +1 @@
-gzip compressed data, was "biocolabsdk-1.0.1.tar", max compression
+gzip compressed data, was "biocolabsdk-1.0.2.tar", max compression
```

## Comparing `biocolabsdk-1.0.1.tar` & `biocolabsdk-1.0.2.tar`

### file list

```diff
@@ -1,5 +1,8 @@
--rw-r--r--   0        0        0     7633 2023-05-15 08:46:28.928703 biocolabsdk-1.0.1/README.md
--rw-r--r--   0        0        0       64 2023-04-12 06:09:13.000000 biocolabsdk-1.0.1/biocolabsdk/__init__.py
--rw-r--r--   0        0        0     4855 2023-04-14 08:48:52.000000 biocolabsdk-1.0.1/biocolabsdk/connector.py
--rw-r--r--   0        0        0      437 2023-05-15 08:50:43.998456 biocolabsdk-1.0.1/pyproject.toml
--rw-r--r--   0        0        0     8290 1970-01-01 00:00:00.000000 biocolabsdk-1.0.1/PKG-INFO
+-rw-r--r--   0        0        0     6774 2023-05-16 11:52:44.788427 biocolabsdk-1.0.2/README.md
+-rw-r--r--   0        0        0     6148 2023-04-13 03:18:11.944270 biocolabsdk-1.0.2/biocolabsdk/.DS_Store
+-rw-r--r--   0        0        0       92 2023-05-16 11:52:44.788427 biocolabsdk-1.0.2/biocolabsdk/__init__.py
+-rw-r--r--   0        0        0     4189 2023-05-16 11:52:44.788427 biocolabsdk-1.0.2/biocolabsdk/connector.py
+-rw-r--r--   0        0        0      756 2023-05-16 11:52:44.788427 biocolabsdk-1.0.2/biocolabsdk/identity.py
+-rw-r--r--   0        0        0     1305 2023-05-16 11:52:44.788427 biocolabsdk-1.0.2/biocolabsdk/util.py
+-rw-r--r--   0        0        0      437 2023-05-16 11:52:44.788427 biocolabsdk-1.0.2/pyproject.toml
+-rw-r--r--   0        0        0     7431 1970-01-01 00:00:00.000000 biocolabsdk-1.0.2/PKG-INFO
```

### Comparing `biocolabsdk-1.0.1/README.md` & `biocolabsdk-1.0.2/README.md`

 * *Files 16% similar despite different names*

```diff
@@ -1,478 +1,424 @@
-00000000: 2323 2031 2e20 5573 6167 6520 6269 6f63  ## 1. Usage bioc
-00000010: 6f6c 6162 7364 6b3a 0a2a 2a54 6865 2070  olabsdk:.**The p
-00000020: 6163 6b61 6765 206f 6e6c 7920 616c 6c6f  ackage only allo
-00000030: 7773 2064 6174 6120 7375 626d 6973 7369  ws data submissi
-00000040: 6f6e 2076 6961 2042 696f 436f 6c61 6220  on via BioColab 
-00000050: 7072 6976 6174 6520 7365 7276 6572 2e20  private server. 
-00000060: 506c 6561 7365 2063 6f6e 6669 6775 7265  Please configure
-00000070: 2079 6f75 7220 746f 6b65 6e73 2069 6e20   your tokens in 
-00000080: 7468 6520 6055 7365 7220 5365 7474 696e  the `User Settin
-00000090: 6773 6020 7061 6765 2e2a 2a0a 2323 2320  gs` page.**.### 
-000000a0: 312e 312e 2054 6573 7420 7468 6520 636f  1.1. Test the co
-000000b0: 6e6e 6563 7469 6f6e 3a0a 6060 600a 2320  nnection:.```.# 
-000000c0: 6578 616d 706c 652e 7079 0a0a 6672 6f6d  example.py..from
-000000d0: 2062 696f 636f 6c61 6273 646b 2e63 6f6e   biocolabsdk.con
-000000e0: 6e65 6374 6f72 2069 6d70 6f72 7420 4543  nector import EC
-000000f0: 6f6e 6e65 6374 6f72 0a0a 636f 6e6e 6563  onnector..connec
-00000100: 746f 7220 3d20 4543 6f6e 6e65 6374 6f72  tor = EConnector
-00000110: 280a 2020 686f 7374 3d22 6874 7470 733a  (.  host="https:
-00000120: 2f2f 796f 7572 636f 6d70 616e 792f 7432  //yourcompany/t2
-00000130: 645f 696e 6465 785f 746f 6f6c 2f2c 0a20  d_index_tool/,. 
-00000140: 2074 6f6b 656e 3d22 3c69 6e70 7574 2079   token="<input y
-00000150: 6f75 7220 746f 6b65 6e20 6865 7265 3e22  our token here>"
-00000160: 0a29 0a0a 636f 6e6e 6563 746f 722e 6765  .)..connector.ge
-00000170: 745f 6262 726f 7773 6572 7828 292e 7465  t_bbrowserx().te
-00000180: 7374 5f63 6f6e 6e65 6374 696f 6e28 290a  st_connection().
-00000190: 6060 600a 0a45 7861 6d70 6c65 206f 7574  ```..Example out
-000001a0: 7075 743a 0a0a 6060 600a 436f 6e6e 6563  put:..```.Connec
-000001b0: 7469 6e67 2074 6f20 686f 7374 2061 7420  ting to host at 
-000001c0: 6874 7470 733a 2f2f 796f 7572 636f 6d70  https://yourcomp
-000001d0: 616e 792f 7432 645f 696e 6465 785f 746f  any/t2d_index_to
-000001e0: 6f6c 2f61 7069 2f76 312f 7465 7374 5f63  ol/api/v1/test_c
-000001f0: 6f6e 6e65 6374 696f 6e0a 436f 6e6e 6563  onnection.Connec
-00000200: 7469 6f6e 2073 7563 6365 7373 6675 6c0a  tion successful.
-00000210: 6060 600a 0a23 2323 2031 2e32 2e20 4765  ```..### 1.2. Ge
-00000220: 7420 7573 6572 2067 726f 7570 7320 6176  t user groups av
-00000230: 6169 6c61 626c 6520 666f 7220 796f 7572  ailable for your
-00000240: 2074 6f6b 656e 3a0a 6060 600a 2320 6578   token:.```.# ex
-00000250: 616d 706c 652e 7079 0a0a 6672 6f6d 2062  ample.py..from b
-00000260: 696f 636f 6c61 6273 646b 2e63 6f6e 6e65  iocolabsdk.conne
-00000270: 6374 6f72 2069 6d70 6f72 7420 4543 6f6e  ctor import ECon
-00000280: 6e65 6374 6f72 0a0a 636f 6e6e 6563 746f  nector..connecto
-00000290: 7220 3d20 4543 6f6e 6e65 6374 6f72 280a  r = EConnector(.
-000002a0: 2020 686f 7374 3d22 6874 7470 733a 2f2f    host="https://
-000002b0: 796f 7572 636f 6d70 616e 792f 7432 645f  yourcompany/t2d_
-000002c0: 696e 6465 785f 746f 6f6c 2f2c 0a20 2074  index_tool/,.  t
-000002d0: 6f6b 656e 3d22 3c69 6e70 7574 2079 6f75  oken="<input you
-000002e0: 7220 746f 6b65 6e20 6865 7265 3e22 0a29  r token here>".)
-000002f0: 0a0a 7573 6572 5f67 726f 7570 7320 3d20  ..user_groups = 
-00000300: 636f 6e6e 6563 746f 722e 6765 745f 6262  connector.get_bb
-00000310: 726f 7773 6572 7828 292e 6765 745f 7573  rowserx().get_us
-00000320: 6572 5f67 726f 7570 7328 290a 7072 696e  er_groups().prin
-00000330: 7428 7573 6572 5f67 726f 7570 7329 0a60  t(user_groups).`
-00000340: 6060 0a0a 4578 616d 706c 6520 6f75 7470  ``..Example outp
-00000350: 7574 3a0a 0a60 6060 0a5b 7b27 6964 273a  ut:..```.[{'id':
-00000360: 2027 616c 6c5f 6d65 6d62 6572 7327 2c20   'all_members', 
-00000370: 276e 616d 6527 3a20 2741 6c6c 206d 656d  'name': 'All mem
-00000380: 6265 7273 277d 2c20 7b27 6964 273a 2027  bers'}, {'id': '
-00000390: 7065 7273 6f6e 616c 272c 2027 6e61 6d65  personal', 'name
-000003a0: 273a 2027 5065 7273 6f6e 616c 2077 6f72  ': 'Personal wor
-000003b0: 6b73 7061 6365 277d 5d0a 6060 600a 0a23  kspace'}].```..#
-000003c0: 2323 2031 2e33 2e20 5375 626d 6974 2068  ## 1.3. Submit h
-000003d0: 3561 6420 2873 6361 6e70 7920 6f62 6a65  5ad (scanpy obje
-000003e0: 6374 293a 0a60 6060 0a23 2065 7861 6d70  ct):.```.# examp
-000003f0: 6c65 2e70 790a 6672 6f6d 2062 696f 636f  le.py.from bioco
-00000400: 6c61 6273 646b 2e63 6f6e 6e65 6374 6f72  labsdk.connector
-00000410: 2069 6d70 6f72 7420 4543 6f6e 6e65 6374   import EConnect
-00000420: 6f72 0a66 726f 6d20 6269 6f74 7572 696e  or.from bioturin
-00000430: 675f 636f 6e6e 6563 746f 722e 7479 7069  g_connector.typi
-00000440: 6e67 2069 6d70 6f72 7420 496e 7075 744d  ng import InputM
-00000450: 6174 7269 7854 7970 650a 6672 6f6d 2062  atrixType.from b
-00000460: 696f 7475 7269 6e67 5f63 6f6e 6e65 6374  ioturing_connect
-00000470: 6f72 2e74 7970 696e 6720 696d 706f 7274  or.typing import
-00000480: 2053 7065 6369 6573 0a0a 636f 6e6e 6563   Species..connec
-00000490: 746f 7220 3d20 4543 6f6e 6e65 6374 6f72  tor = EConnector
-000004a0: 280a 2020 686f 7374 3d22 6874 7470 733a  (.  host="https:
-000004b0: 2f2f 796f 7572 636f 6d70 616e 792f 7432  //yourcompany/t2
-000004c0: 645f 696e 6465 785f 746f 6f6c 2f2c 0a20  d_index_tool/,. 
-000004d0: 2074 6f6b 656e 3d22 3c69 6e70 7574 2079   token="<input y
-000004e0: 6f75 7220 746f 6b65 6e20 6865 7265 3e22  our token here>"
-000004f0: 0a29 0a0a 2320 4361 6c6c 2074 6869 7320  .)..# Call this 
-00000500: 6675 6e63 7469 6f6e 2066 6972 7374 2074  function first t
-00000510: 6f20 6765 7420 6176 6169 6c61 626c 6520  o get available 
-00000520: 6772 6f75 7073 2061 6e64 2074 6865 6972  groups and their
-00000530: 2069 642e 0a75 7365 725f 6772 6f75 7073   id..user_groups
-00000540: 203d 2063 6f6e 6e65 6374 6f72 2e67 6574   = connector.get
-00000550: 5f62 6272 6f77 7365 7278 2829 2e67 6574  _bbrowserx().get
-00000560: 5f75 7365 725f 6772 6f75 7073 2829 0a23  _user_groups().#
-00000570: 2045 7861 6d70 6c65 3a20 7573 6572 5f67   Example: user_g
-00000580: 726f 7570 7320 6973 206e 6f77 205b 7b27  roups is now [{'
-00000590: 6964 273a 2027 616c 6c5f 6d65 6d62 6572  id': 'all_member
-000005a0: 7327 2c20 276e 616d 6527 3a20 2741 6c6c  s', 'name': 'All
-000005b0: 206d 656d 6265 7273 277d 2c20 7b27 6964   members'}, {'id
-000005c0: 273a 2027 7065 7273 6f6e 616c 272c 2027  ': 'personal', '
-000005d0: 6e61 6d65 273a 2027 5065 7273 6f6e 616c  name': 'Personal
-000005e0: 2077 6f72 6b73 7061 6365 277d 5d0a 0a0a   workspace'}]...
-000005f0: 2320 5375 626d 6974 7469 6e67 2074 6865  # Submitting the
-00000600: 2073 6361 6e70 7920 6f62 6a65 6374 3a0a   scanpy object:.
-00000610: 636f 6e6e 6563 746f 722e 7375 626d 6974  connector.submit
-00000620: 5f68 3561 6428 0a20 2067 726f 7570 5f69  _h5ad(.  group_i
-00000630: 643d 2770 6572 736f 6e61 6c27 2c0a 2020  d='personal',.  
-00000640: 7374 7564 795f 7333 5f6b 6579 733d 5b27  study_s3_keys=['
-00000650: 4753 4531 3238 3232 332e 6835 6164 275d  GSE128223.h5ad']
-00000660: 2c0a 2020 7374 7564 795f 6964 3d27 4753  ,.  study_id='GS
-00000670: 4531 3238 3232 3327 2c0a 2020 6e61 6d65  E128223',.  name
-00000680: 3d27 5468 6973 2069 7320 6d79 2066 6972  ='This is my fir
-00000690: 7374 2073 7475 6479 272c 0a20 2061 7574  st study',.  aut
-000006a0: 686f 7273 3d5b 2748 7579 204e 6775 7965  hors=['Huy Nguye
-000006b0: 6e27 5d2c 0a20 2073 7065 6369 6573 3d53  n'],.  species=S
-000006c0: 7065 6369 6573 2e48 554d 414e 2e76 616c  pecies.HUMAN.val
-000006d0: 7565 2c0a 2020 696e 7075 745f 6d61 7472  ue,.  input_matr
-000006e0: 6978 5f74 7970 653d 496e 7075 744d 6174  ix_type=InputMat
-000006f0: 7269 7854 7970 652e 5241 572e 7661 6c75  rixType.RAW.valu
-00000700: 650a 290a 0a23 2045 7861 6d70 6c65 206f  e.)..# Example o
-00000710: 7574 7075 743a 0a3e 205b 3230 3232 2d31  utput:.> [2022-1
-00000720: 302d 3130 2030 313a 3033 5d20 5761 6974  0-10 01:03] Wait
-00000730: 696e 6720 696e 2071 7565 7565 0a3e 205b  ing in queue.> [
-00000740: 3230 3232 2d31 302d 3130 2030 313a 3033  2022-10-10 01:03
-00000750: 5d20 446f 776e 6c6f 6164 696e 6720 4753  ] Downloading GS
-00000760: 4531 3238 3232 332e 6835 6164 2066 726f  E128223.h5ad fro
-00000770: 6d20 7333 3a20 3236 322e 3120 4b42 202f  m s3: 262.1 KB /
-00000780: 2034 3332 2e38 204d 420a 3e20 5b32 3032   432.8 MB.> [202
-00000790: 322d 3130 2d31 3020 3031 3a30 335d 2046  2-10-10 01:03] F
-000007a0: 696c 6520 646f 776e 6c6f 6164 6564 0a3e  ile downloaded.>
-000007b0: 205b 3230 3232 2d31 302d 3130 2030 313a   [2022-10-10 01:
-000007c0: 3033 5d20 5265 6164 696e 6720 6261 7463  03] Reading batc
-000007d0: 683a 2047 5345 3132 3832 3233 2e68 3561  h: GSE128223.h5a
-000007e0: 640a 3e20 5b32 3032 322d 3130 2d31 3020  d.> [2022-10-10 
-000007f0: 3031 3a30 335d 2050 7265 7072 6f63 6573  01:03] Preproces
-00000800: 7369 6e67 2065 7870 7265 7373 696f 6e20  sing expression 
-00000810: 6d61 7472 6978 3a20 3139 3132 3120 6365  matrix: 19121 ce
-00000820: 6c6c 7320 7820 3633 3831 3320 6765 6e65  lls x 63813 gene
-00000830: 730a 3e20 5b32 3032 322d 3130 2d31 3020  s.> [2022-10-10 
-00000840: 3031 3a30 335d 2046 696c 7465 7265 643a  01:03] Filtered:
-00000850: 2031 3931 3231 2063 656c 6c73 2072 656d   19121 cells rem
-00000860: 6169 6e0a 3e20 5b32 3032 322d 3130 2d31  ain.> [2022-10-1
-00000870: 3020 3031 3a30 335d 2053 7461 7274 2070  0 01:03] Start p
-00000880: 726f 6365 7373 696e 6720 7374 7564 790a  rocessing study.
-00000890: 3e20 5b32 3032 322d 3130 2d31 3020 3031  > [2022-10-10 01
-000008a0: 3a30 335d 204e 6f72 6d61 6c69 7a69 6e67  :03] Normalizing
-000008b0: 2065 7870 7265 7373 696f 6e20 6d61 7472   expression matr
-000008c0: 6978 0a3e 205b 3230 3232 2d31 302d 3130  ix.> [2022-10-10
-000008d0: 2030 313a 3033 5d20 5275 6e6e 696e 6720   01:03] Running 
-000008e0: 5043 410a 3e20 5b32 3032 322d 3130 2d31  PCA.> [2022-10-1
-000008f0: 3020 3031 3a30 335d 2052 756e 6e69 6e67  0 01:03] Running
-00000900: 206b 4e4e 0a3e 205b 3230 3232 2d31 302d   kNN.> [2022-10-
-00000910: 3130 2030 313a 3033 5d20 5275 6e6e 696e  10 01:03] Runnin
-00000920: 6720 7370 6563 7472 616c 2065 6d62 6564  g spectral embed
-00000930: 6469 6e67 0a3e 205b 3230 3232 2d31 302d  ding.> [2022-10-
-00000940: 3130 2030 313a 3033 5d20 5275 6e6e 696e  10 01:03] Runnin
-00000950: 6720 7665 6e69 6365 2062 696e 6172 697a  g venice binariz
-00000960: 6572 0a3e 205b 3230 3232 2d31 302d 3130  er.> [2022-10-10
-00000970: 2030 313a 3034 5d20 5275 6e6e 696e 6720   01:04] Running 
-00000980: 742d 534e 450a 3e20 5b32 3032 322d 3130  t-SNE.> [2022-10
-00000990: 2d31 3020 3031 3a30 345d 2053 7475 6479  -10 01:04] Study
-000009a0: 2077 6173 2073 7563 6365 7373 6675 6c6c   was successfull
-000009b0: 7920 7375 626d 6974 7465 640a 3e20 5b32  y submitted.> [2
-000009c0: 3032 322d 3130 2d31 3020 3031 3a30 345d  022-10-10 01:04]
-000009d0: 2044 4f4e 4520 2121 210a 3e20 5374 7564   DONE !!!.> Stud
-000009e0: 7920 7375 626d 6974 7465 6420 7375 6363  y submitted succ
-000009f0: 6573 7366 756c 6c79 210a 6060 600a 4176  essfully!.```.Av
-00000a00: 6169 6c61 626c 6520 7061 7261 6d65 7465  ailable paramete
-00000a10: 7273 2066 6f72 2060 7375 626d 6974 5f68  rs for `submit_h
-00000a20: 3561 6460 2066 756e 6374 696f 6e3a 0a60  5ad` function:.`
-00000a30: 6060 0a67 726f 7570 5f69 643a 2073 7472  ``.group_id: str
-00000a40: 0a20 2049 4420 6f66 2074 6865 2067 726f  .  ID of the gro
-00000a50: 7570 2074 6f20 7375 626d 6974 2074 6865  up to submit the
-00000a60: 2064 6174 6120 746f 2e0a 0a73 7475 6479   data to...study
-00000a70: 5f73 335f 6b65 7973 3a20 4c69 7374 5b73  _s3_keys: List[s
-00000a80: 7472 5d0a 2020 4c69 7374 206f 6620 7468  tr].  List of th
-00000a90: 6520 7333 206b 6579 206f 6620 7468 6520  e s3 key of the 
-00000aa0: 7374 7564 6965 732e 0a0a 7374 7564 795f  studies...study_
-00000ab0: 6964 3a20 7374 722c 2064 6566 6175 6c74  id: str, default
-00000ac0: 3d4e 6f6e 650a 2020 5374 7564 7920 4944  =None.  Study ID
-00000ad0: 2c20 6966 206e 6f20 7661 6c75 6520 6973  , if no value is
-00000ae0: 2073 7065 6369 6669 6564 2c20 7573 6520   specified, use 
-00000af0: 6120 7261 6e64 6f6d 2075 7569 6476 3420  a random uuidv4 
-00000b00: 7374 7269 6e67 0a0a 6e61 6d65 3a20 7374  string..name: st
-00000b10: 722c 2064 6566 6175 6c74 3d27 546f 2062  r, default='To b
-00000b20: 6520 6465 7461 696c 6564 270a 2020 4e61  e detailed'.  Na
-00000b30: 6d65 206f 6620 7468 6520 7374 7564 792e  me of the study.
-00000b40: 0a0a 6175 7468 6f72 733a 204c 6973 745b  ..authors: List[
-00000b50: 7374 725d 2c20 6465 6661 756c 743d 5b5d  str], default=[]
-00000b60: 0a20 2041 7574 686f 7273 206f 6620 7468  .  Authors of th
-00000b70: 6520 7374 7564 792e 0a0a 6162 7374 7261  e study...abstra
-00000b80: 6374 3a20 7374 722c 2064 6566 6175 6c74  ct: str, default
-00000b90: 3d27 270a 2020 4162 7374 7261 6374 206f  =''.  Abstract o
-00000ba0: 6620 7468 6520 7374 7564 792e 0a0a 7370  f the study...sp
-00000bb0: 6563 6965 733a 2073 7472 2c20 6465 6661  ecies: str, defa
-00000bc0: 756c 743d 2768 756d 616e 270a 2020 5370  ult='human'.  Sp
-00000bd0: 6563 6965 7320 6f66 2074 6865 2073 7475  ecies of the stu
-00000be0: 6479 2e20 4361 6e20 6265 3a20 2a2a 6269  dy. Can be: **bi
-00000bf0: 6f74 7572 696e 675f 636f 6e6e 6563 746f  oturing_connecto
-00000c00: 722e 7479 7069 6e67 2e53 7065 6369 6573  r.typing.Species
-00000c10: 2e48 554d 414e 2e76 616c 7565 2a2a 206f  .HUMAN.value** o
-00000c20: 7220 2a2a 6269 6f74 7572 696e 675f 636f  r **bioturing_co
-00000c30: 6e6e 6563 746f 722e 7479 7069 6e67 2e53  nnector.typing.S
-00000c40: 7065 6369 6573 2e4d 4f55 5345 2e76 616c  pecies.MOUSE.val
-00000c50: 7565 2a2a 206f 7220 2a2a 6269 6f74 7572  ue** or **biotur
-00000c60: 696e 675f 636f 6e6e 6563 746f 722e 7479  ing_connector.ty
-00000c70: 7069 6e67 2e53 7065 6369 6573 2e4e 4f4e  ping.Species.NON
-00000c80: 5f48 554d 414e 5f50 5249 4d41 5445 2e76  _HUMAN_PRIMATE.v
-00000c90: 616c 7565 2a2a 0a0a 696e 7075 745f 6d61  alue**..input_ma
-00000ca0: 7472 6978 5f74 7970 653a 2073 7472 2c20  trix_type: str, 
-00000cb0: 6465 6661 756c 743d 2772 6177 270a 2020  default='raw'.  
-00000cc0: 4966 2074 6865 2076 616c 7565 206f 6620  If the value of 
-00000cd0: 7468 6973 2069 6e70 7574 2069 7320 2a2a  this input is **
-00000ce0: 6269 6f74 7572 696e 675f 636f 6e6e 6563  bioturing_connec
-00000cf0: 746f 722e 7479 7069 6e67 2e49 6e70 7574  tor.typing.Input
-00000d00: 4d61 7472 6978 5479 7065 2e4e 4f52 4d41  MatrixType.NORMA
-00000d10: 4c49 5a45 442e 7661 6c75 652a 2a2c 0a20  LIZED.value**,. 
-00000d20: 2074 6865 6e20 7468 6520 736f 6674 7761   then the softwa
-00000d30: 7265 2077 696c 6c0a 2020 7573 6520 736c  re will.  use sl
-00000d40: 6f74 2027 5827 2066 726f 6d20 7468 6520  ot 'X' from the 
-00000d50: 7363 616e 7079 206f 626a 6563 7420 616e  scanpy object an
-00000d60: 6420 646f 6573 206e 6f74 2061 7070 6c79  d does not apply
-00000d70: 206e 6f72 6d61 6c69 7a61 7469 6f6e 2e0a   normalization..
-00000d80: 2020 4966 2074 6865 2076 616c 7565 206f    If the value o
-00000d90: 6620 7468 6973 2069 6e70 7574 2069 7320  f this input is 
-00000da0: 2a2a 6269 6f74 7572 696e 675f 636f 6e6e  **bioturing_conn
-00000db0: 6563 746f 722e 7479 7069 6e67 2e49 6e70  ector.typing.Inp
-00000dc0: 7574 4d61 7472 6978 5479 7065 2e52 4157  utMatrixType.RAW
-00000dd0: 2e76 616c 7565 2a2a 2c74 6865 6e20 7468  .value**,then th
-00000de0: 6520 736f 6674 7761 7265 2077 696c 6c0a  e software will.
-00000df0: 2020 7573 6520 736c 6f74 2027 7261 772e    use slot 'raw.
-00000e00: 5827 2066 726f 6d20 7468 6573 6361 6e70  X' from thescanp
-00000e10: 7920 6f62 6a65 6374 2061 6e64 2061 7070  y object and app
-00000e20: 6c79 206c 6f67 2d6e 6f72 6d61 6c69 7a61  ly log-normaliza
-00000e30: 7469 6f6e 2e0a 0a6d 696e 5f63 6f75 6e74  tion...min_count
-00000e40: 733a 2069 6e74 2c20 6465 6661 756c 743d  s: int, default=
-00000e50: 4e6f 6e65 0a20 204d 696e 696d 756d 206e  None.  Minimum n
-00000e60: 756d 6265 7220 6f66 2063 6f75 6e74 7320  umber of counts 
-00000e70: 7265 7175 6972 6564 0a20 2066 6f72 2061  required.  for a
-00000e80: 2063 656c 6c20 746f 2070 6173 7320 6669   cell to pass fi
-00000e90: 6c74 6572 696e 672e 0a0a 6d69 6e5f 6765  ltering...min_ge
-00000ea0: 6e65 733a 2069 6e74 2c20 6465 6661 756c  nes: int, defaul
-00000eb0: 743d 4e6f 6e65 0a20 204d 696e 696d 756d  t=None.  Minimum
-00000ec0: 206e 756d 6265 7220 6f66 2067 656e 6573   number of genes
-00000ed0: 2065 7870 7265 7373 6564 2072 6571 7569   expressed requi
-00000ee0: 7265 640a 2020 666f 7220 6120 6365 6c6c  red.  for a cell
-00000ef0: 2074 6f20 7061 7373 2066 696c 7465 7269   to pass filteri
-00000f00: 6e67 2e0a 0a6d 6178 5f63 6f75 6e74 733a  ng...max_counts:
-00000f10: 2069 6e74 2c20 6465 6661 756c 743d 4e6f   int, default=No
-00000f20: 6e65 0a20 204d 6178 696d 756d 206e 756d  ne.  Maximum num
-00000f30: 6265 7220 6f66 2063 6f75 6e74 7320 7265  ber of counts re
-00000f40: 7175 6972 6564 0a20 2066 6f72 2061 2063  quired.  for a c
-00000f50: 656c 6c20 746f 2070 6173 7320 6669 6c74  ell to pass filt
-00000f60: 6572 696e 672e 0a0a 6d61 785f 6765 6e65  ering...max_gene
-00000f70: 733a 2069 6e74 2c20 6465 6661 756c 743d  s: int, default=
-00000f80: 4e6f 6e65 0a20 204d 6178 696d 756d 206e  None.  Maximum n
-00000f90: 756d 6265 7220 6f66 2067 656e 6573 2065  umber of genes e
-00000fa0: 7870 7265 7373 6564 2072 6571 7569 7265  xpressed require
-00000fb0: 640a 2020 666f 7220 6120 6365 6c6c 2074  d.  for a cell t
-00000fc0: 6f20 7061 7373 2066 696c 7465 7269 6e67  o pass filtering
-00000fd0: 2e0a 0a6d 745f 7065 7263 656e 7461 6765  ...mt_percentage
-00000fe0: 3a20 556e 696f 6e5b 696e 742c 2066 6c6f  : Union[int, flo
-00000ff0: 6174 5d2c 2064 6566 6175 6c74 3d4e 6f6e  at], default=Non
-00001000: 650a 2020 4d61 7869 6d75 6d20 6e75 6d62  e.  Maximum numb
-00001010: 6572 206f 6620 6d69 746f 6368 6f6e 6472  er of mitochondr
-00001020: 6961 2067 656e 6573 2070 6572 6365 6e74  ia genes percent
-00001030: 6167 650a 2020 7265 7175 6972 6564 2066  age.  required f
-00001040: 6f72 2061 2063 656c 6c20 746f 2070 6173  or a cell to pas
-00001050: 7320 6669 6c74 6572 696e 672e 2052 616e  s filtering. Ran
-00001060: 6769 6e67 2066 726f 6d20 3020 746f 2031  ging from 0 to 1
-00001070: 3030 0a60 6060 0a0a 2323 2032 2e20 5573  00.```..## 2. Us
-00001080: 6167 6520 6269 6f66 6c65 783a 0a0a 2323  age bioflex:..##
-00001090: 2320 4372 6561 7465 2061 2063 6f6e 6e65  # Create a conne
-000010a0: 6374 696f 6e20 7573 696e 6720 6163 6365  ction using acce
-000010b0: 7373 2074 6f6b 656e 3a0a 0a60 6060 7079  ss token:..```py
-000010c0: 7468 6f6e 0a66 726f 6d20 6269 6f63 6f6c  thon.from biocol
-000010d0: 6162 7364 6b2e 636f 6e6e 6563 746f 7220  absdk.connector 
-000010e0: 696d 706f 7274 2045 436f 6e6e 6563 746f  import EConnecto
-000010f0: 720a 0a63 6f6e 6e65 6374 6f72 203d 2045  r..connector = E
-00001100: 436f 6e6e 6563 746f 7228 0a20 2070 7562  Connector(.  pub
-00001110: 6c69 635f 746f 6b65 6e3d 223c 696e 7075  lic_token="<inpu
-00001120: 7420 796f 7572 2074 6f6b 656e 2068 6572  t your token her
-00001130: 653e 220a 290a 6060 600a 0a23 2323 204c  e>".).```..### L
-00001140: 6973 7420 6176 6169 6c61 626c 6520 6461  ist available da
-00001150: 7461 6261 7365 733a 0a0a 6060 6070 7974  tabases:..```pyt
-00001160: 686f 6e0a 6461 7461 6261 7365 7320 3d20  hon.databases = 
-00001170: 636f 6e6e 6563 7469 6f6e 2e67 6574 5f62  connection.get_b
-00001180: 696f 666c 6578 2829 2e64 6174 6162 6173  ioflex().databas
-00001190: 6573 2829 0a60 6060 0a3e 6060 600a 3e20  es().```.>```.> 
-000011a0: 5b44 6174 6142 6173 6528 6964 3d22 3530  [DataBase(id="50
-000011b0: 3130 6337 6435 3733 6165 3466 6632 6239  10c7d573ae4ff2b9
-000011c0: 3639 3134 3232 6239 3961 6132 6364 222c  691422b99aa2cd",
-000011d0: 0a3e 2020 2020 2020 2020 2020 206e 616d  .>           nam
-000011e0: 653d 2242 696f 5475 7269 6e67 2064 6174  e="BioTuring dat
-000011f0: 6162 6173 6522 2c73 7065 6369 6573 3d22  abase",species="
-00001200: 6875 6d61 6e22 2c76 6572 7369 6f6e 3d31  human",version=1
-00001210: 292c 0a3e 2044 6174 6142 6173 6528 6964  ),.> DataBase(id
-00001220: 3d22 3530 3130 6337 6435 3733 6165 3466  ="5010c7d573ae4f
-00001230: 6632 6239 3639 3134 3232 6239 3961 6132  f2b9691422b99aa2
-00001240: 6364 222c 0a3e 2020 2020 2020 2020 2020  cd",.>          
-00001250: 206e 616d 653d 2242 696f 5475 7269 6e67   name="BioTuring
-00001260: 2064 6174 6162 6173 6522 2c73 7065 6369   database",speci
-00001270: 6573 3d22 6875 6d61 6e22 2c76 6572 7369  es="human",versi
-00001280: 6f6e 3d32 292c 0a3e 2044 6174 6142 6173  on=2),.> DataBas
-00001290: 6528 6964 3d22 3530 3130 6337 6435 3733  e(id="5010c7d573
-000012a0: 6165 3466 6632 6239 3639 3134 3232 6239  ae4ff2b9691422b9
-000012b0: 3961 6132 6364 222c 0a3e 2020 2020 2020  9aa2cd",.>      
-000012c0: 2020 2020 206e 616d 653d 2242 696f 5475       name="BioTu
-000012d0: 7269 6e67 2064 6174 6162 6173 6522 2c73  ring database",s
-000012e0: 7065 6369 6573 3d22 6875 6d61 6e22 2c76  pecies="human",v
-000012f0: 6572 7369 6f6e 3d33 295d 0a0a 2323 2320  ersion=3)]..### 
-00001300: 4765 7420 6461 7461 6261 7365 2063 656c  Get database cel
-00001310: 6c20 7479 7065 7320 6765 6e65 2065 7870  l types gene exp
-00001320: 7265 7373 696f 6e20 7375 6d6d 6172 790a  ression summary.
-00001330: 0a60 6060 7079 7468 6f6e 0a64 6174 6162  .```python.datab
-00001340: 6173 6520 3d20 6461 7461 6261 7365 735b  ase = databases[
-00001350: 325d 0a64 6174 6162 6173 652e 6765 745f  2].database.get_
-00001360: 6365 6c6c 7479 7065 735f 6578 7072 6573  celltypes_expres
-00001370: 7369 6f6e 5f73 756d 6d61 7279 285b 2743  sion_summary(['C
-00001380: 4433 4427 2c20 2743 4433 4527 5d29 0a60  D3D', 'CD3E']).`
-00001390: 6060 0a3e 6060 600a 3e20 7b27 4344 3344  ``.>```.> {'CD3D
-000013a0: 273a 205b 5375 6d6d 6172 7928 6e61 6d65  ': [Summary(name
-000013b0: 3d22 4220 6365 6c6c 222c 7375 6d3d 3730  ="B cell",sum=70
-000013c0: 3731 3038 3837 342e 302c 6d65 616e 3d34  7108874.0,mean=4
-000013d0: 3139 322e 3730 3936 2c72 6174 653d 302e  192.7096,rate=0.
-000013e0: 3033 352c 636f 756e 743d 3136 3836 3532  035,count=168652
-000013f0: 2e30 2c74 6f74 616c 3d34 3831 3239 3637  .0,total=4812967
-00001400: 292c 0a3e 2009 5375 6d6d 6172 7928 6e61  ),.> .Summary(na
-00001410: 6d65 3d22 4344 342d 706f 7369 7469 7665  me="CD4-positive
-00001420: 2c20 616c 7068 612d 6265 7461 2054 2063  , alpha-beta T c
-00001430: 656c 6c22 2c73 756d 3d39 3438 3939 3837  ell",sum=9489987
-00001440: 3434 322e 302c 6d65 616e 3d34 3635 372e  442.0,mean=4657.
-00001450: 3536 3139 2c72 6174 653d 302e 3532 3833  5619,rate=0.5283
-00001460: 2c63 6f75 6e74 3d32 3033 3735 3434 2e30  ,count=2037544.0
-00001470: 2c74 6f74 616c 3d33 3835 3635 3930 292c  ,total=3856590),
-00001480: 0a3e 2009 2e2e 2e0a 3e20 0953 756d 6d61  .> .....> .Summa
-00001490: 7279 286e 616d 653d 2263 6f72 6e65 616c  ry(name="corneal
-000014a0: 2070 726f 6765 6e69 746f 7222 2c73 756d   progenitor",sum
-000014b0: 3d30 2e30 2c6d 6561 6e3d 302e 302c 7261  =0.0,mean=0.0,ra
-000014c0: 7465 3d30 2e30 2c63 6f75 6e74 3d30 2e30  te=0.0,count=0.0
-000014d0: 2c74 6f74 616c 3d33 3937 3329 2c0a 3e20  ,total=3973),.> 
-000014e0: 0953 756d 6d61 7279 286e 616d 653d 226e  .Summary(name="n
-000014f0: 7563 6c65 7573 2070 756c 706f 7375 7320  ucleus pulposus 
-00001500: 7072 6f67 656e 6974 6f72 2063 656c 6c22  progenitor cell"
-00001510: 2c73 756d 3d30 2e30 2c6d 6561 6e3d 302e  ,sum=0.0,mean=0.
-00001520: 302c 7261 7465 3d30 2e30 2c63 6f75 6e74  0,rate=0.0,count
-00001530: 3d30 2e30 2c74 6f74 616c 3d32 3331 3029  =0.0,total=2310)
-00001540: 5d7d 0a0a 0a23 2323 2043 7265 6174 6520  ]}...### Create 
-00001550: 7374 7564 7920 696e 7374 616e 6365 2c20  study instance, 
-00001560: 7573 696e 6720 7374 7564 7920 6861 7368  using study hash
-00001570: 2049 4420 6672 6f6d 205b 4269 6f54 7572   ID from [BioTur
-00001580: 696e 6720 7374 7564 6965 735d 2868 7474  ing studies](htt
-00001590: 7073 3a2f 2f74 616c 6b32 6461 7461 2e62  ps://talk2data.b
-000015a0: 696f 7475 7269 6e67 2e63 6f6d 2f73 7475  ioturing.com/stu
-000015b0: 6469 6573 2f29 3a0a 0a60 6060 7079 7468  dies/):..```pyth
-000015c0: 6f6e 0a73 7475 6479 203d 2064 6174 6162  on.study = datab
-000015d0: 6173 652e 6765 745f 7374 7564 7928 2747  ase.get_study('G
-000015e0: 5345 3936 3538 335f 6261 7463 6832 2729  SE96583_batch2')
-000015f0: 0a73 7475 6479 0a60 6060 0a3e 6060 600a  .study.```.>```.
-00001600: 3e20 5374 7564 7928 6964 3d22 4753 4539  > Study(id="GSE9
-00001610: 3635 3833 5f62 6174 6368 3222 2c0a 3e20  6583_batch2",.> 
-00001620: 2020 2020 2020 7469 746c 653d 224d 756c        title="Mul
-00001630: 7469 706c 6578 6564 2064 726f 706c 6574  tiplexed droplet
-00001640: 2073 696e 676c 652d 6365 6c6c 2052 4e41   single-cell RNA
-00001650: 2d73 6571 7565 6e63 696e 6720 7573 696e  -sequencing usin
-00001660: 6720 6e61 7475 7261 6c20 6765 6e65 7469  g natural geneti
-00001670: 6320 7661 7269 6174 696f 6e20 2842 6174  c variation (Bat
-00001680: 6368 2032 2922 2c0a 3e20 2020 2020 2020  ch 2)",.>       
-00001690: 7265 6665 7265 6e63 653d 2268 7474 7073  reference="https
-000016a0: 3a2f 2f77 7777 2e6e 6174 7572 652e 636f  ://www.nature.co
-000016b0: 6d2f 6172 7469 636c 6573 2f6e 6274 2e34  m/articles/nbt.4
-000016c0: 3034 3222 290a 0a23 2323 2054 616b 6520  042")..### Take 
-000016d0: 6120 7065 656b 2061 7420 7374 7564 7920  a peek at study 
-000016e0: 6d65 7461 6461 7461 3a0a 0a60 6060 7079  metadata:..```py
-000016f0: 7468 6f6e 0a73 7475 6479 2e6d 6574 616c  thon.study.metal
-00001700: 6973 740a 6060 600a 3e60 6060 0a3e 205b  ist.```.>```.> [
-00001710: 4d65 7461 6461 7461 2869 643d 302c 6e61  Metadata(id=0,na
-00001720: 6d65 3d22 4e75 6d62 6572 206f 6620 6d52  me="Number of mR
-00001730: 4e41 2074 7261 6e73 6372 6970 7473 222c  NA transcripts",
-00001740: 7479 7065 3d22 4e75 6d65 7269 6322 292c  type="Numeric"),
-00001750: 0a3e 2020 4d65 7461 6461 7461 2869 643d  .>  Metadata(id=
-00001760: 312c 6e61 6d65 3d22 4e75 6d62 6572 206f  1,name="Number o
-00001770: 6620 6765 6e65 7322 2c74 7970 653d 224e  f genes",type="N
-00001780: 756d 6572 6963 2229 2c0a 3e20 204d 6574  umeric"),.>  Met
-00001790: 6164 6174 6128 6964 3d32 2c6e 616d 653d  adata(id=2,name=
-000017a0: 2242 6174 6368 2069 6422 2c74 7970 653d  "Batch id",type=
-000017b0: 2243 6174 6567 6f72 7922 292c 0a3e 2020  "Category"),.>  
-000017c0: 4d65 7461 6461 7461 2869 643d 332c 6e61  Metadata(id=3,na
-000017d0: 6d65 3d22 5374 696d 756c 6174 696f 6e22  me="Stimulation"
-000017e0: 2c74 7970 653d 2243 6174 6567 6f72 7922  ,type="Category"
-000017f0: 292c 0a3e 2020 4d65 7461 6461 7461 2869  ),.>  Metadata(i
-00001800: 643d 342c 6e61 6d65 3d22 4175 7468 6f72  d=4,name="Author
-00001810: 2773 2063 656c 6c20 7479 7065 222c 7479  's cell type",ty
-00001820: 7065 3d22 4361 7465 676f 7279 2229 5d0a  pe="Category")].
-00001830: 0a23 2323 2046 6574 6368 2061 2073 7475  .### Fetch a stu
-00001840: 6479 206d 6574 6164 6174 613a 0a0a 6060  dy metadata:..``
-00001850: 6070 7974 686f 6e0a 6d65 7461 6461 7461  `python.metadata
-00001860: 203d 2073 7475 6479 2e6d 6574 616c 6973   = study.metalis
-00001870: 745b 345d 0a6d 6574 6164 6174 610a 6060  t[4].metadata.``
-00001880: 600a 3e60 6060 0a3e 4d65 7461 6461 7461  `.>```.>Metadata
-00001890: 2869 643d 342c 6e61 6d65 3d22 4175 7468  (id=4,name="Auth
-000018a0: 6f72 2773 2063 656c 6c20 7479 7065 222c  or's cell type",
-000018b0: 7479 7065 3d22 4361 7465 676f 7279 2229  type="Category")
-000018c0: 0a60 6060 7079 7468 6f6e 0a6d 6574 6164  .```python.metad
-000018d0: 6174 612e 6665 7463 6828 290a 6d65 7461  ata.fetch().meta
-000018e0: 6461 7461 2e76 616c 7565 730a 6060 600a  data.values.```.
-000018f0: 3e60 6060 0a3e 2061 7272 6179 285b 2743  >```.> array(['C
-00001900: 4438 2054 2063 656c 6c73 272c 2027 4465  D8 T cells', 'De
-00001910: 6e64 7269 7469 6320 6365 6c6c 7327 2c20  ndritic cells', 
-00001920: 2743 4434 2054 2063 656c 6c73 272c 202e  'CD4 T cells', .
-00001930: 2e2e 2c0a 3e20 2020 2020 2020 2027 4344  ..,.>        'CD
-00001940: 3820 5420 6365 6c6c 7327 2c20 2742 2063  8 T cells', 'B c
-00001950: 656c 6c73 272c 2027 4344 3420 5420 6365  ells', 'CD4 T ce
-00001960: 6c6c 7327 5d2c 2064 7479 7065 3d27 3c55  lls'], dtype='<U
-00001970: 3137 2729 0a0a 2323 2320 5175 6572 7920  17')..### Query 
-00001980: 6765 6e65 733a 0a0a 6060 6070 7974 686f  genes:..```pytho
-00001990: 6e0a 6578 705f 6d74 7820 3d20 7374 7564  n.exp_mtx = stud
-000019a0: 792e 7175 6572 795f 6765 6e65 7328 5b27  y.query_genes(['
-000019b0: 4344 3344 272c 2027 4344 3345 275d 2c20  CD3D', 'CD3E'], 
-000019c0: 6269 6f66 6c65 782e 554e 4954 5f4c 4f47  bioflex.UNIT_LOG
-000019d0: 4e4f 524d 290a 6578 705f 6d74 780a 6060  NORM).exp_mtx.``
-000019e0: 600a 3e60 6060 0a3e 203c 3239 3036 3578  `.>```.> <29065x
-000019f0: 3220 7370 6172 7365 206d 6174 7269 7820  2 sparse matrix 
-00001a00: 6f66 2074 7970 6520 273c 636c 6173 7320  of type '<class 
-00001a10: 276e 756d 7079 2e66 6c6f 6174 3332 273e  'numpy.float32'>
-00001a20: 270a 3e20 2020 2020 7769 7468 2031 3534  '.>     with 154
-00001a30: 3932 2073 746f 7265 6420 656c 656d 656e  92 stored elemen
-00001a40: 7473 2069 6e20 436f 6d70 7265 7373 6564  ts in Compressed
-00001a50: 2053 7061 7273 6520 436f 6c75 6d6e 2066   Sparse Column f
-00001a60: 6f72 6d61 743e 0a0a 2323 2320 4765 7420  ormat>..### Get 
-00001a70: 7374 7564 7920 6261 7263 6f64 6573 3a0a  study barcodes:.
-00001a80: 0a60 6060 7079 7468 6f6e 0a73 7475 6479  .```python.study
-00001a90: 2e62 6172 636f 6465 7328 290a 6060 600a  .barcodes().```.
-00001aa0: 3e60 6060 0a3e 205b 2747 534d 3235 3630  >```.> ['GSM2560
-00001ab0: 3234 395f 4141 4143 4154 4143 4341 4147  249_AAACATACCAAG
-00001ac0: 4354 2d31 272c 0a3e 2020 2747 534d 3235  CT-1',.>  'GSM25
-00001ad0: 3630 3234 395f 4141 4143 4154 4143 4343  60249_AAACATACCC
-00001ae0: 4354 4143 2d31 272c 0a3e 2020 2e2e 2e0a  CTAC-1',.>  ....
-00001af0: 3e20 2027 4753 4d32 3536 3032 3439 5f41  >  'GSM2560249_A
-00001b00: 4154 5447 5447 4154 5443 4143 542d 3127  ATTGTGATTCACT-1'
-00001b10: 2c0a 3e20 2027 4753 4d32 3536 3032 3439  ,.>  'GSM2560249
-00001b20: 5f41 4154 5447 5447 4154 5454 4347 542d  _AATTGTGATTTCGT-
-00001b30: 3127 2c0a 3e20 202e 2e2e 5d0a 0a23 2323  1',.>  ...]..###
-00001b40: 2047 6574 2073 7475 6479 2066 6561 7475   Get study featu
-00001b50: 7265 733a 0a0a 6060 6070 7974 686f 6e0a  res:..```python.
-00001b60: 7374 7564 792e 6665 6174 7572 6573 2829  study.features()
-00001b70: 0a60 6060 0a3e 6060 600a 3e20 5b27 3553  .```.>```.> ['5S
-00001b80: 5f52 524e 4127 2c0a 3e20 2027 355f 3853  _RRNA',.>  '5_8S
-00001b90: 5f52 524e 4127 2c0a 3e20 202e 2e2e 0a3e  _RRNA',.>  ....>
-00001ba0: 2020 2741 4330 3036 3237 3327 2c0a 3e20    'AC006273',.> 
-00001bb0: 2027 4143 3030 3632 3737 272c 0a3e 2020   'AC006277',.>  
-00001bc0: 2e2e 2e5d 0a0a 2323 2320 4765 7420 7374  ...]..### Get st
-00001bd0: 7564 7920 6675 6c6c 206d 6174 7269 783a  udy full matrix:
-00001be0: 0a0a 6060 6070 7974 686f 6e0a 7374 7564  ..```python.stud
-00001bf0: 792e 6d61 7472 6978 2862 696f 666c 6578  y.matrix(bioflex
-00001c00: 2e55 4e49 545f 4c4f 474e 4f52 4d29 0a60  .UNIT_LOGNORM).`
-00001c10: 6060 0a3e 6060 600a 3e20 3c32 3930 3635  ``.>```.> <29065
-00001c20: 7836 3436 3432 2073 7061 7273 6520 6d61  x64642 sparse ma
-00001c30: 7472 6978 206f 6620 7479 7065 2027 3c63  trix of type '<c
-00001c40: 6c61 7373 2027 6e75 6d70 792e 666c 6f61  lass 'numpy.floa
-00001c50: 7433 3227 3e27 0a3e 2009 7769 7468 2031  t32'>'.> .with 1
-00001c60: 3735 3730 3733 3920 7374 6f72 6564 2065  7570739 stored e
-00001c70: 6c65 6d65 6e74 7320 696e 2043 6f6d 7072  lements in Compr
-00001c80: 6573 7365 6420 5370 6172 7365 2043 6f6c  essed Sparse Col
-00001c90: 756d 6e20 666f 726d 6174 3e0a 0a23 2323  umn format>..###
-00001ca0: 2045 7870 6f72 7420 5374 7564 793a 0a0a   Export Study:..
-00001cb0: 6060 6070 7974 686f 6e0a 7374 7564 792e  ```python.study.
-00001cc0: 6578 706f 7274 5f73 7475 6479 2862 696f  export_study(bio
-00001cd0: 666c 6578 2e45 5850 4f52 545f 4835 4144  flex.EXPORT_H5AD
-00001ce0: 290a 6060 600a 3e60 6060 0a3e 7b27 646f  ).```.>```.>{'do
-00001cf0: 776e 6c6f 6164 5f6c 696e 6b27 3a20 2768  wnload_link': 'h
-00001d00: 7474 7073 3a2f 2f74 616c 6b32 6461 7461  ttps://talk2data
-00001d10: 2e62 696f 7475 7269 6e67 2e63 6f6d 2f61  .bioturing.com/a
-00001d20: 7069 2f65 7870 6f72 742f 6131 3030 3362  pi/export/a1003b
-00001d30: 6164 3364 6431 3436 6232 3863 3762 6461  ad3dd146b28c7bda
-00001d40: 3931 3361 3266 6333 6630 272c 0a3e 2027  913a2fc3f0',.> '
-00001d50: 7374 7564 795f 6861 7368 5f69 6427 3a20  study_hash_id': 
-00001d60: 2747 5345 3936 3538 335f 6261 7463 6832  'GSE96583_batch2
-00001d70: 277d 0a0a 2d2d 2d2d 0a46 6f72 2066 7572  '}..----.For fur
-00001d80: 7468 6572 2069 6e66 6f72 6d61 7469 6f6e  ther information
-00001d90: 2070 6c65 6173 6520 6368 6563 6b20 7468   please check th
-00001da0: 6520 5b64 6f63 756d 656e 7461 7469 6f6e  e [documentation
-00001db0: 5d28 6874 7470 733a 2f2f 636f 6c61 622e  ](https://colab.
-00001dc0: 6269 6f74 7572 696e 672e 636f 6d2f 292e  bioturing.com/).
-00001dd0: 0a                                       .
+00000000: 2320 5573 6167 6520 6269 6f63 6f6c 6162  # Usage biocolab
+00000010: 7364 6b3a 0a0a 2323 2031 2e20 5573 6167  sdk:..## 1. Usag
+00000020: 6520 6262 726f 7773 6572 7820 696e 2062  e bbrowserx in b
+00000030: 696f 636f 6c61 6273 646b 3a0a 0a2a 2a54  iocolabsdk:..**T
+00000040: 6865 2070 6163 6b61 6765 206f 6e6c 7920  he package only 
+00000050: 616c 6c6f 7773 2064 6174 6120 7375 626d  allows data subm
+00000060: 6973 7369 6f6e 2076 6961 2042 696f 436f  ission via BioCo
+00000070: 6c61 6220 7072 6976 6174 6520 7365 7276  lab private serv
+00000080: 6572 2e20 506c 6561 7365 2063 6f6e 6669  er. Please confi
+00000090: 6775 7265 2079 6f75 7220 746f 6b65 6e73  gure your tokens
+000000a0: 2069 6e20 7468 6520 6055 7365 7220 5365   in the `User Se
+000000b0: 7474 696e 6773 6020 7061 6765 2e2a 2a0a  ttings` page.**.
+000000c0: 0a3c 2f62 723e 0a3c 696d 6720 616c 743d  .</br>.<img alt=
+000000d0: 2273 6574 7469 6e67 2220 7372 633d 2268  "setting" src="h
+000000e0: 7474 7073 3a2f 2f63 646e 2e62 696f 7475  ttps://cdn.biotu
+000000f0: 7269 6e67 2e63 6f6d 2f64 6f63 756d 656e  ring.com/documen
+00000100: 7461 7469 6f6e 2f6d 642f 7573 6572 2d73  tation/md/user-s
+00000110: 6574 7469 6e67 2e70 6e67 2220 7769 6474  etting.png" widt
+00000120: 683d 2231 3030 2522 3e0a 3c2f 6272 3e0a  h="100%">.</br>.
+00000130: 0a23 2323 2031 2e31 2e20 5465 7374 2074  .### 1.1. Test t
+00000140: 6865 2063 6f6e 6e65 6374 696f 6e20 746f  he connection to
+00000150: 2042 4272 6f77 7365 7258 2070 7269 7661   BBrowserX priva
+00000160: 7465 2073 6572 7665 723a 0a60 6060 0a23  te server:.```.#
+00000170: 2065 7861 6d70 6c65 2e70 790a 0a66 726f   example.py..fro
+00000180: 6d20 6269 6f63 6f6c 6162 7364 6b2e 636f  m biocolabsdk.co
+00000190: 6e6e 6563 746f 7220 696d 706f 7274 2045  nnector import E
+000001a0: 436f 6e6e 6563 746f 720a 0a63 6f6e 6e65  Connector..conne
+000001b0: 6374 6f72 203d 2045 436f 6e6e 6563 746f  ctor = EConnecto
+000001c0: 7228 0a20 2070 7269 7661 7465 5f68 6f73  r(.  private_hos
+000001d0: 743d 2268 7474 7073 3a2f 2f79 6f75 7263  t="https://yourc
+000001e0: 6f6d 7061 6e79 2f74 3264 5f69 6e64 6578  ompany/t2d_index
+000001f0: 5f74 6f6f 6c2f 2c0a 2020 7072 6976 6174  _tool/,.  privat
+00000200: 655f 746f 6b65 6e3d 223c 696e 7075 7420  e_token="<input 
+00000210: 796f 7572 2074 6f6b 656e 2068 6572 653e  your token here>
+00000220: 220a 290a 0a63 6f6e 6e65 6374 6f72 2e67  ".)..connector.g
+00000230: 6574 5f62 6272 6f77 7365 7278 2829 2e74  et_bbrowserx().t
+00000240: 6573 745f 636f 6e6e 6563 7469 6f6e 2829  est_connection()
+00000250: 0a60 6060 0a0a 4578 616d 706c 6520 6f75  .```..Example ou
+00000260: 7470 7574 3a0a 0a60 6060 0a43 6f6e 6e65  tput:..```.Conne
+00000270: 6374 696e 6720 746f 2068 6f73 7420 6174  cting to host at
+00000280: 2068 7474 7073 3a2f 2f79 6f75 7263 6f6d   https://yourcom
+00000290: 7061 6e79 2f74 3264 5f69 6e64 6578 5f74  pany/t2d_index_t
+000002a0: 6f6f 6c2f 6170 692f 7631 2f74 6573 745f  ool/api/v1/test_
+000002b0: 636f 6e6e 6563 7469 6f6e 0a43 6f6e 6e65  connection.Conne
+000002c0: 6374 696f 6e20 7375 6363 6573 7366 756c  ction successful
+000002d0: 0a60 6060 0a0a 2323 2320 312e 322e 2047  .```..### 1.2. G
+000002e0: 6574 2075 7365 7220 6772 6f75 7073 2061  et user groups a
+000002f0: 7661 696c 6162 6c65 2066 6f72 2079 6f75  vailable for you
+00000300: 7220 746f 6b65 6e20 696e 2042 4272 6f77  r token in BBrow
+00000310: 7365 7258 2070 7269 7661 7465 2073 6572  serX private ser
+00000320: 7665 723a 0a60 6060 0a23 2065 7861 6d70  ver:.```.# examp
+00000330: 6c65 2e70 790a 0a66 726f 6d20 6269 6f63  le.py..from bioc
+00000340: 6f6c 6162 7364 6b2e 636f 6e6e 6563 746f  olabsdk.connecto
+00000350: 7220 696d 706f 7274 2045 436f 6e6e 6563  r import EConnec
+00000360: 746f 720a 0a63 6f6e 6e65 6374 6f72 203d  tor..connector =
+00000370: 2045 436f 6e6e 6563 746f 7228 0a20 2070   EConnector(.  p
+00000380: 7269 7661 7465 5f68 6f73 743d 2268 7474  rivate_host="htt
+00000390: 7073 3a2f 2f79 6f75 7263 6f6d 7061 6e79  ps://yourcompany
+000003a0: 2f74 3264 5f69 6e64 6578 5f74 6f6f 6c2f  /t2d_index_tool/
+000003b0: 2c0a 2020 7072 6976 6174 655f 746f 6b65  ,.  private_toke
+000003c0: 6e3d 223c 696e 7075 7420 796f 7572 2074  n="<input your t
+000003d0: 6f6b 656e 2068 6572 653e 220a 290a 0a75  oken here>".)..u
+000003e0: 7365 725f 6772 6f75 7073 203d 2063 6f6e  ser_groups = con
+000003f0: 6e65 6374 6f72 2e67 6574 5f62 6272 6f77  nector.get_bbrow
+00000400: 7365 7278 2829 2e67 6574 5f75 7365 725f  serx().get_user_
+00000410: 6772 6f75 7073 2829 0a70 7269 6e74 2875  groups().print(u
+00000420: 7365 725f 6772 6f75 7073 290a 6060 600a  ser_groups).```.
+00000430: 0a45 7861 6d70 6c65 206f 7574 7075 743a  .Example output:
+00000440: 0a0a 6060 600a 5b7b 2769 6427 3a20 2761  ..```.[{'id': 'a
+00000450: 6c6c 5f6d 656d 6265 7273 272c 2027 6e61  ll_members', 'na
+00000460: 6d65 273a 2027 416c 6c20 6d65 6d62 6572  me': 'All member
+00000470: 7327 7d2c 207b 2769 6427 3a20 2770 6572  s'}, {'id': 'per
+00000480: 736f 6e61 6c27 2c20 276e 616d 6527 3a20  sonal', 'name': 
+00000490: 2750 6572 736f 6e61 6c20 776f 726b 7370  'Personal worksp
+000004a0: 6163 6527 7d5d 0a60 6060 0a0a 2323 2320  ace'}].```..### 
+000004b0: 312e 332e 2053 7562 6d69 7420 796f 7572  1.3. Submit your
+000004c0: 2064 6174 6120 746f 2042 4272 6f77 7365   data to BBrowse
+000004d0: 7258 2070 7269 7661 7465 2073 6572 7665  rX private serve
+000004e0: 723a 0a60 6060 0a23 2065 7861 6d70 6c65  r:.```.# example
+000004f0: 2e70 790a 6672 6f6d 2062 696f 636f 6c61  .py.from biocola
+00000500: 6273 646b 2e63 6f6e 6e65 6374 6f72 2069  bsdk.connector i
+00000510: 6d70 6f72 7420 4543 6f6e 6e65 6374 6f72  mport EConnector
+00000520: 0a66 726f 6d20 6269 6f74 7572 696e 675f  .from bioturing_
+00000530: 636f 6e6e 6563 746f 722e 7479 7069 6e67  connector.typing
+00000540: 2069 6d70 6f72 7420 496e 7075 744d 6174   import InputMat
+00000550: 7269 7854 7970 650a 6672 6f6d 2062 696f  rixType.from bio
+00000560: 7475 7269 6e67 5f63 6f6e 6e65 6374 6f72  turing_connector
+00000570: 2e74 7970 696e 6720 696d 706f 7274 2053  .typing import S
+00000580: 7065 6369 6573 0a66 726f 6d20 6269 6f74  pecies.from biot
+00000590: 7572 696e 675f 636f 6e6e 6563 746f 722e  uring_connector.
+000005a0: 7479 7069 6e67 2069 6d70 6f72 7420 5374  typing import St
+000005b0: 7564 7954 7970 650a 0a63 6f6e 6e65 6374  udyType..connect
+000005c0: 6f72 203d 2045 436f 6e6e 6563 746f 7228  or = EConnector(
+000005d0: 0a20 2070 7269 7661 7465 5f68 6f73 743d  .  private_host=
+000005e0: 2268 7474 7073 3a2f 2f79 6f75 7263 6f6d  "https://yourcom
+000005f0: 7061 6e79 2f74 3264 5f69 6e64 6578 5f74  pany/t2d_index_t
+00000600: 6f6f 6c2f 2c0a 2020 7072 6976 6174 655f  ool/,.  private_
+00000610: 746f 6b65 6e3d 223c 696e 7075 7420 796f  token="<input yo
+00000620: 7572 2074 6f6b 656e 2068 6572 653e 220a  ur token here>".
+00000630: 290a 0a23 2043 616c 6c20 7468 6973 2066  )..# Call this f
+00000640: 756e 6374 696f 6e20 6669 7273 7420 746f  unction first to
+00000650: 2067 6574 2061 7661 696c 6162 6c65 2067   get available g
+00000660: 726f 7570 7320 616e 6420 7468 6569 7220  roups and their 
+00000670: 6964 2e0a 7573 6572 5f67 726f 7570 7320  id..user_groups 
+00000680: 3d20 636f 6e6e 6563 746f 722e 6765 745f  = connector.get_
+00000690: 6262 726f 7773 6572 7828 292e 6765 745f  bbrowserx().get_
+000006a0: 7573 6572 5f67 726f 7570 7328 290a 2320  user_groups().# 
+000006b0: 4578 616d 706c 653a 2075 7365 725f 6772  Example: user_gr
+000006c0: 6f75 7073 2069 7320 6e6f 7720 5b7b 2769  oups is now [{'i
+000006d0: 6427 3a20 2761 6c6c 5f6d 656d 6265 7273  d': 'all_members
+000006e0: 272c 2027 6e61 6d65 273a 2027 416c 6c20  ', 'name': 'All 
+000006f0: 6d65 6d62 6572 7327 7d2c 207b 2769 6427  members'}, {'id'
+00000700: 3a20 2770 6572 736f 6e61 6c27 2c20 276e  : 'personal', 'n
+00000710: 616d 6527 3a20 2750 6572 736f 6e61 6c20  ame': 'Personal 
+00000720: 776f 726b 7370 6163 6527 7d5d 0a0a 2320  workspace'}]..# 
+00000730: 5375 626d 6974 7469 6e67 2074 6865 2073  Submitting the s
+00000740: 6361 6e70 7920 6f62 6a65 6374 2066 726f  canpy object fro
+00000750: 6d20 7333 3a0a 6261 7463 685f 696e 666f  m s3:.batch_info
+00000760: 203d 205b 7b0a 2020 2020 276d 6174 7269   = [{.    'matri
+00000770: 7827 3a20 2773 335f 7061 7468 2f47 5345  x': 's3_path/GSE
+00000780: 3132 3832 3233 5f31 2e68 3561 6427 2c0a  128223_1.h5ad',.
+00000790: 7d2c 207b 0a20 2020 2027 6d61 7472 6978  }, {.    'matrix
+000007a0: 273a 2027 7333 5f70 6174 682f 4753 4531  ': 's3_path/GSE1
+000007b0: 3238 3232 335f 322e 6835 6164 272c 0a7d  28223_2.h5ad',.}
+000007c0: 5d0a 0a63 6f6e 6e65 6374 6f72 2e67 6574  ]..connector.get
+000007d0: 5f62 6272 6f77 7365 7278 2829 2e73 7562  _bbrowserx().sub
+000007e0: 6d69 745f 7374 7564 795f 6672 6f6d 5f73  mit_study_from_s
+000007f0: 3328 0a20 2067 726f 7570 5f69 643d 2770  3(.  group_id='p
+00000800: 6572 736f 6e61 6c27 2c0a 2020 6261 7463  ersonal',.  batc
+00000810: 685f 696e 666f 3d62 6174 6368 5f69 6e66  h_info=batch_inf
+00000820: 6f2c 0a20 2073 7475 6479 5f69 643d 2747  o,.  study_id='G
+00000830: 5345 3132 3832 3233 272c 0a20 206e 616d  SE128223',.  nam
+00000840: 653d 2754 6869 7320 6973 206d 7920 6669  e='This is my fi
+00000850: 7273 7420 7374 7564 7927 2c0a 2020 6175  rst study',.  au
+00000860: 7468 6f72 733d 5b27 4875 7920 4e67 7579  thors=['Huy Nguy
+00000870: 656e 275d 2c0a 2020 7370 6563 6965 733d  en'],.  species=
+00000880: 5370 6563 6965 732e 4855 4d41 4e2e 7661  Species.HUMAN.va
+00000890: 6c75 652c 0a20 2069 6e70 7574 5f6d 6174  lue,.  input_mat
+000008a0: 7269 785f 7479 7065 3d49 6e70 7574 4d61  rix_type=InputMa
+000008b0: 7472 6978 5479 7065 2e52 4157 2e76 616c  trixType.RAW.val
+000008c0: 7565 2c0a 2020 7374 7564 795f 7479 7065  ue,.  study_type
+000008d0: 3d53 7475 6479 5479 7065 2e48 3541 442e  =StudyType.H5AD.
+000008e0: 7661 6c75 650a 290a 0a23 2053 7562 6d69  value.)..# Submi
+000008f0: 7474 696e 6720 7468 6520 7363 616e 7079  tting the scanpy
+00000900: 206f 626a 6563 7420 6672 6f6d 206c 6f63   object from loc
+00000910: 616c 206d 6163 6869 6e65 3a0a 6261 7463  al machine:.batc
+00000920: 685f 696e 666f 203d 205b 7b0a 2020 2020  h_info = [{.    
+00000930: 276d 6174 7269 7827 3a20 276c 6f63 616c  'matrix': 'local
+00000940: 5f70 6174 682f 4753 4531 3238 3232 335f  _path/GSE128223_
+00000950: 312e 6835 6164 272c 0a7d 2c20 7b0a 2020  1.h5ad',.}, {.  
+00000960: 2020 276d 6174 7269 7827 3a20 276c 6f63    'matrix': 'loc
+00000970: 616c 5f70 6174 682f 4753 4531 3238 3232  al_path/GSE12822
+00000980: 335f 322e 6835 6164 272c 0a7d 5d0a 0a63  3_2.h5ad',.}]..c
+00000990: 6f6e 6e65 6374 6f72 2e67 6574 5f62 6272  onnector.get_bbr
+000009a0: 6f77 7365 7278 2829 2e73 7562 6d69 745f  owserx().submit_
+000009b0: 7374 7564 795f 6672 6f6d 5f6c 6f63 616c  study_from_local
+000009c0: 280a 2020 6772 6f75 705f 6964 3d27 7065  (.  group_id='pe
+000009d0: 7273 6f6e 616c 272c 0a20 2062 6174 6368  rsonal',.  batch
+000009e0: 5f69 6e66 6f3d 6261 7463 685f 696e 666f  _info=batch_info
+000009f0: 2c0a 2020 7374 7564 795f 6964 3d27 4753  ,.  study_id='GS
+00000a00: 4531 3238 3232 3327 2c0a 2020 6e61 6d65  E128223',.  name
+00000a10: 3d27 5468 6973 2069 7320 6d79 2066 6972  ='This is my fir
+00000a20: 7374 2073 7475 6479 272c 0a20 2061 7574  st study',.  aut
+00000a30: 686f 7273 3d5b 2748 7579 204e 6775 7965  hors=['Huy Nguye
+00000a40: 6e27 5d2c 0a20 2073 7065 6369 6573 3d53  n'],.  species=S
+00000a50: 7065 6369 6573 2e48 554d 414e 2e76 616c  pecies.HUMAN.val
+00000a60: 7565 2c0a 2020 696e 7075 745f 6d61 7472  ue,.  input_matr
+00000a70: 6978 5f74 7970 653d 496e 7075 744d 6174  ix_type=InputMat
+00000a80: 7269 7854 7970 652e 5241 572e 7661 6c75  rixType.RAW.valu
+00000a90: 652c 0a20 2073 7475 6479 5f74 7970 653d  e,.  study_type=
+00000aa0: 5374 7564 7954 7970 652e 4835 4144 2e76  StudyType.H5AD.v
+00000ab0: 616c 7565 0a29 0a0a 2323 2032 2e20 5573  alue.)..## 2. Us
+00000ac0: 6167 6520 7461 6c6b 3264 6174 6120 696e  age talk2data in
+00000ad0: 2062 696f 636f 6c61 6273 646b 3a0a 0a60   biocolabsdk:..`
+00000ae0: 6060 0a23 2065 7861 6d70 6c65 2e70 790a  ``.# example.py.
+00000af0: 0a66 726f 6d20 6269 6f63 6f6c 6162 7364  .from biocolabsd
+00000b00: 6b2e 636f 6e6e 6563 746f 7220 696d 706f  k.connector impo
+00000b10: 7274 2045 436f 6e6e 6563 746f 720a 0a63  rt EConnector..c
+00000b20: 6f6e 6e65 6374 6f72 203d 2045 436f 6e6e  onnector = EConn
+00000b30: 6563 746f 7228 0a20 2070 7562 6c69 635f  ector(.  public_
+00000b40: 746f 6b65 6e3d 223c 696e 7075 7420 796f  token="<input yo
+00000b50: 7572 2074 6f6b 656e 2068 6572 653e 220a  ur token here>".
+00000b60: 290a 0a63 6f6e 6e65 6374 6f72 2e67 6574  )..connector.get
+00000b70: 5f74 616c 6b32 6461 7461 2829 2e74 6573  _talk2data().tes
+00000b80: 745f 636f 6e6e 6563 7469 6f6e 2829 0a60  t_connection().`
+00000b90: 6060 0a0a 4578 616d 706c 6520 6f75 7470  ``..Example outp
+00000ba0: 7574 3a0a 0a60 6060 0a43 6f6e 6e65 6374  ut:..```.Connect
+00000bb0: 696e 6720 746f 2068 6f73 7420 6174 2068  ing to host at h
+00000bc0: 7474 7073 3a2f 2f74 616c 6b32 6461 7461  ttps://talk2data
+00000bd0: 2e62 696f 7475 7269 6e67 2e63 6f6d 2f74  .bioturing.com/t
+00000be0: 3264 5f69 6e64 6578 5f74 6f6f 6c2f 6170  2d_index_tool/ap
+00000bf0: 692f 7631 2f74 6573 745f 636f 6e6e 6563  i/v1/test_connec
+00000c00: 7469 6f6e 0a43 6f6e 6e65 6374 696f 6e20  tion.Connection 
+00000c10: 7375 6363 6573 7366 756c 0a60 6060 0a0a  successful.```..
+00000c20: 596f 7520 6361 6e20 7574 696c 697a 6520  You can utilize 
+00000c30: 616c 6c20 7468 6520 6665 6174 7572 6573  all the features
+00000c40: 2069 6e20 6120 7369 6d69 6c61 7220 7761   in a similar wa
+00000c50: 7920 746f 2061 6363 6573 7369 6e67 2061  y to accessing a
+00000c60: 2042 4272 6f73 6572 5820 7072 6976 6174   BBroserX privat
+00000c70: 6520 7365 7276 6572 2e0a 0a23 2320 332e  e server...## 3.
+00000c80: 2055 7361 6765 2062 696f 666c 6578 2069   Usage bioflex i
+00000c90: 6e20 6269 6f63 6f6c 6162 7364 6b3a 0a0a  n biocolabsdk:..
+00000ca0: 596f 7520 6361 6e20 6765 7420 6120 4269  You can get a Bi
+00000cb0: 6f66 6c65 7820 746f 6b65 6e20 6174 3a20  oflex token at: 
+00000cc0: 0a3c 6120 6872 6566 3d22 6874 7470 733a  .<a href="https:
+00000cd0: 2f2f 636f 6c61 622e 6269 6f74 7572 696e  //colab.bioturin
+00000ce0: 672e 636f 6d2f 7265 7175 6573 742f 6269  g.com/request/bi
+00000cf0: 6f66 6c65 782f 746f 6b65 6e22 3e68 7474  oflex/token">htt
+00000d00: 7073 3a2f 2f63 6f6c 6162 2e62 696f 7475  ps://colab.biotu
+00000d10: 7269 6e67 2e63 6f6d 2f72 6571 7565 7374  ring.com/request
+00000d20: 2f62 696f 666c 6578 2f74 6f6b 656e 3c2f  /bioflex/token</
+00000d30: 613e 0a0a 2323 2320 4372 6561 7465 2061  a>..### Create a
+00000d40: 2063 6f6e 6e65 6374 696f 6e20 7573 696e   connection usin
+00000d50: 6720 6163 6365 7373 2074 6f6b 656e 3a0a  g access token:.
+00000d60: 0a60 6060 7079 7468 6f6e 0a66 726f 6d20  .```python.from 
+00000d70: 6269 6f63 6f6c 6162 7364 6b2e 636f 6e6e  biocolabsdk.conn
+00000d80: 6563 746f 7220 696d 706f 7274 2045 436f  ector import ECo
+00000d90: 6e6e 6563 746f 720a 0a63 6f6e 6e65 6374  nnector..connect
+00000da0: 6f72 203d 2045 436f 6e6e 6563 746f 7228  or = EConnector(
+00000db0: 0a20 2062 696f 666c 6578 5f74 6f6b 656e  .  bioflex_token
+00000dc0: 3d22 3c69 6e70 7574 2079 6f75 7220 746f  ="<input your to
+00000dd0: 6b65 6e20 6865 7265 3e22 0a29 0a60 6060  ken here>".).```
+00000de0: 0a0a 2323 2320 4c69 7374 2061 7661 696c  ..### List avail
+00000df0: 6162 6c65 2064 6174 6162 6173 6573 3a0a  able databases:.
+00000e00: 0a60 6060 7079 7468 6f6e 0a64 6174 6162  .```python.datab
+00000e10: 6173 6573 203d 2063 6f6e 6e65 6374 6f72  ases = connector
+00000e20: 2e67 6574 5f62 696f 666c 6578 2829 2e64  .get_bioflex().d
+00000e30: 6174 6162 6173 6573 2829 0a60 6060 0a3e  atabases().```.>
+00000e40: 6060 600a 3e20 5b44 6174 6142 6173 6528  ```.> [DataBase(
+00000e50: 6964 3d22 3530 3130 6337 6435 3733 6165  id="5010c7d573ae
+00000e60: 3466 6632 6239 3639 3134 3232 6239 3961  4ff2b9691422b99a
+00000e70: 6132 6364 222c 0a3e 2020 2020 2020 2020  a2cd",.>        
+00000e80: 2020 206e 616d 653d 2242 696f 5475 7269     name="BioTuri
+00000e90: 6e67 2064 6174 6162 6173 6522 2c73 7065  ng database",spe
+00000ea0: 6369 6573 3d22 6875 6d61 6e22 2c76 6572  cies="human",ver
+00000eb0: 7369 6f6e 3d31 292c 0a3e 2044 6174 6142  sion=1),.> DataB
+00000ec0: 6173 6528 6964 3d22 3530 3130 6337 6435  ase(id="5010c7d5
+00000ed0: 3733 6165 3466 6632 6239 3639 3134 3232  73ae4ff2b9691422
+00000ee0: 6239 3961 6132 6364 222c 0a3e 2020 2020  b99aa2cd",.>    
+00000ef0: 2020 2020 2020 206e 616d 653d 2242 696f         name="Bio
+00000f00: 5475 7269 6e67 2064 6174 6162 6173 6522  Turing database"
+00000f10: 2c73 7065 6369 6573 3d22 6875 6d61 6e22  ,species="human"
+00000f20: 2c76 6572 7369 6f6e 3d32 292c 0a3e 2044  ,version=2),.> D
+00000f30: 6174 6142 6173 6528 6964 3d22 3530 3130  ataBase(id="5010
+00000f40: 6337 6435 3733 6165 3466 6632 6239 3639  c7d573ae4ff2b969
+00000f50: 3134 3232 6239 3961 6132 6364 222c 0a3e  1422b99aa2cd",.>
+00000f60: 2020 2020 2020 2020 2020 206e 616d 653d             name=
+00000f70: 2242 696f 5475 7269 6e67 2064 6174 6162  "BioTuring datab
+00000f80: 6173 6522 2c73 7065 6369 6573 3d22 6875  ase",species="hu
+00000f90: 6d61 6e22 2c76 6572 7369 6f6e 3d33 295d  man",version=3)]
+00000fa0: 0a0a 2323 2320 4765 7420 6461 7461 6261  ..### Get databa
+00000fb0: 7365 2063 656c 6c20 7479 7065 7320 6765  se cell types ge
+00000fc0: 6e65 2065 7870 7265 7373 696f 6e20 7375  ne expression su
+00000fd0: 6d6d 6172 790a 0a60 6060 7079 7468 6f6e  mmary..```python
+00000fe0: 0a64 6174 6162 6173 6520 3d20 6461 7461  .database = data
+00000ff0: 6261 7365 735b 325d 0a64 6174 6162 6173  bases[2].databas
+00001000: 652e 6765 745f 6365 6c6c 7479 7065 735f  e.get_celltypes_
+00001010: 6578 7072 6573 7369 6f6e 5f73 756d 6d61  expression_summa
+00001020: 7279 285b 2743 4433 4427 2c20 2743 4433  ry(['CD3D', 'CD3
+00001030: 4527 5d29 0a60 6060 0a3e 6060 600a 3e20  E']).```.>```.> 
+00001040: 7b27 4344 3344 273a 205b 5375 6d6d 6172  {'CD3D': [Summar
+00001050: 7928 6e61 6d65 3d22 4220 6365 6c6c 222c  y(name="B cell",
+00001060: 7375 6d3d 3730 3731 3038 3837 342e 302c  sum=707108874.0,
+00001070: 6d65 616e 3d34 3139 322e 3730 3936 2c72  mean=4192.7096,r
+00001080: 6174 653d 302e 3033 352c 636f 756e 743d  ate=0.035,count=
+00001090: 3136 3836 3532 2e30 2c74 6f74 616c 3d34  168652.0,total=4
+000010a0: 3831 3239 3637 292c 0a3e 2009 5375 6d6d  812967),.> .Summ
+000010b0: 6172 7928 6e61 6d65 3d22 4344 342d 706f  ary(name="CD4-po
+000010c0: 7369 7469 7665 2c20 616c 7068 612d 6265  sitive, alpha-be
+000010d0: 7461 2054 2063 656c 6c22 2c73 756d 3d39  ta T cell",sum=9
+000010e0: 3438 3939 3837 3434 322e 302c 6d65 616e  489987442.0,mean
+000010f0: 3d34 3635 372e 3536 3139 2c72 6174 653d  =4657.5619,rate=
+00001100: 302e 3532 3833 2c63 6f75 6e74 3d32 3033  0.5283,count=203
+00001110: 3735 3434 2e30 2c74 6f74 616c 3d33 3835  7544.0,total=385
+00001120: 3635 3930 292c 0a3e 2009 2e2e 2e0a 3e20  6590),.> .....> 
+00001130: 0953 756d 6d61 7279 286e 616d 653d 2263  .Summary(name="c
+00001140: 6f72 6e65 616c 2070 726f 6765 6e69 746f  orneal progenito
+00001150: 7222 2c73 756d 3d30 2e30 2c6d 6561 6e3d  r",sum=0.0,mean=
+00001160: 302e 302c 7261 7465 3d30 2e30 2c63 6f75  0.0,rate=0.0,cou
+00001170: 6e74 3d30 2e30 2c74 6f74 616c 3d33 3937  nt=0.0,total=397
+00001180: 3329 2c0a 3e20 0953 756d 6d61 7279 286e  3),.> .Summary(n
+00001190: 616d 653d 226e 7563 6c65 7573 2070 756c  ame="nucleus pul
+000011a0: 706f 7375 7320 7072 6f67 656e 6974 6f72  posus progenitor
+000011b0: 2063 656c 6c22 2c73 756d 3d30 2e30 2c6d   cell",sum=0.0,m
+000011c0: 6561 6e3d 302e 302c 7261 7465 3d30 2e30  ean=0.0,rate=0.0
+000011d0: 2c63 6f75 6e74 3d30 2e30 2c74 6f74 616c  ,count=0.0,total
+000011e0: 3d32 3331 3029 5d7d 0a0a 0a23 2323 2043  =2310)]}...### C
+000011f0: 7265 6174 6520 7374 7564 7920 696e 7374  reate study inst
+00001200: 616e 6365 2c20 7573 696e 6720 7374 7564  ance, using stud
+00001210: 7920 6861 7368 2049 4420 6672 6f6d 205b  y hash ID from [
+00001220: 4269 6f54 7572 696e 6720 7374 7564 6965  BioTuring studie
+00001230: 735d 2868 7474 7073 3a2f 2f74 616c 6b32  s](https://talk2
+00001240: 6461 7461 2e62 696f 7475 7269 6e67 2e63  data.bioturing.c
+00001250: 6f6d 2f73 7475 6469 6573 2f29 3a0a 0a60  om/studies/):..`
+00001260: 6060 7079 7468 6f6e 0a73 7475 6479 203d  ``python.study =
+00001270: 2064 6174 6162 6173 652e 6765 745f 7374   database.get_st
+00001280: 7564 7928 2747 5345 3936 3538 335f 6261  udy('GSE96583_ba
+00001290: 7463 6832 2729 0a73 7475 6479 0a60 6060  tch2').study.```
+000012a0: 0a3e 6060 600a 3e20 5374 7564 7928 6964  .>```.> Study(id
+000012b0: 3d22 4753 4539 3635 3833 5f62 6174 6368  ="GSE96583_batch
+000012c0: 3222 2c0a 3e20 2020 2020 2020 7469 746c  2",.>       titl
+000012d0: 653d 224d 756c 7469 706c 6578 6564 2064  e="Multiplexed d
+000012e0: 726f 706c 6574 2073 696e 676c 652d 6365  roplet single-ce
+000012f0: 6c6c 2052 4e41 2d73 6571 7565 6e63 696e  ll RNA-sequencin
+00001300: 6720 7573 696e 6720 6e61 7475 7261 6c20  g using natural 
+00001310: 6765 6e65 7469 6320 7661 7269 6174 696f  genetic variatio
+00001320: 6e20 2842 6174 6368 2032 2922 2c0a 3e20  n (Batch 2)",.> 
+00001330: 2020 2020 2020 7265 6665 7265 6e63 653d        reference=
+00001340: 2268 7474 7073 3a2f 2f77 7777 2e6e 6174  "https://www.nat
+00001350: 7572 652e 636f 6d2f 6172 7469 636c 6573  ure.com/articles
+00001360: 2f6e 6274 2e34 3034 3222 290a 0a23 2323  /nbt.4042")..###
+00001370: 2054 616b 6520 6120 7065 656b 2061 7420   Take a peek at 
+00001380: 7374 7564 7920 6d65 7461 6461 7461 3a0a  study metadata:.
+00001390: 0a60 6060 7079 7468 6f6e 0a73 7475 6479  .```python.study
+000013a0: 2e6d 6574 616c 6973 740a 6060 600a 3e60  .metalist.```.>`
+000013b0: 6060 0a3e 205b 4d65 7461 6461 7461 2869  ``.> [Metadata(i
+000013c0: 643d 302c 6e61 6d65 3d22 4e75 6d62 6572  d=0,name="Number
+000013d0: 206f 6620 6d52 4e41 2074 7261 6e73 6372   of mRNA transcr
+000013e0: 6970 7473 222c 7479 7065 3d22 4e75 6d65  ipts",type="Nume
+000013f0: 7269 6322 292c 0a3e 2020 4d65 7461 6461  ric"),.>  Metada
+00001400: 7461 2869 643d 312c 6e61 6d65 3d22 4e75  ta(id=1,name="Nu
+00001410: 6d62 6572 206f 6620 6765 6e65 7322 2c74  mber of genes",t
+00001420: 7970 653d 224e 756d 6572 6963 2229 2c0a  ype="Numeric"),.
+00001430: 3e20 204d 6574 6164 6174 6128 6964 3d32  >  Metadata(id=2
+00001440: 2c6e 616d 653d 2242 6174 6368 2069 6422  ,name="Batch id"
+00001450: 2c74 7970 653d 2243 6174 6567 6f72 7922  ,type="Category"
+00001460: 292c 0a3e 2020 4d65 7461 6461 7461 2869  ),.>  Metadata(i
+00001470: 643d 332c 6e61 6d65 3d22 5374 696d 756c  d=3,name="Stimul
+00001480: 6174 696f 6e22 2c74 7970 653d 2243 6174  ation",type="Cat
+00001490: 6567 6f72 7922 292c 0a3e 2020 4d65 7461  egory"),.>  Meta
+000014a0: 6461 7461 2869 643d 342c 6e61 6d65 3d22  data(id=4,name="
+000014b0: 4175 7468 6f72 2773 2063 656c 6c20 7479  Author's cell ty
+000014c0: 7065 222c 7479 7065 3d22 4361 7465 676f  pe",type="Catego
+000014d0: 7279 2229 5d0a 0a23 2323 2046 6574 6368  ry")]..### Fetch
+000014e0: 2061 2073 7475 6479 206d 6574 6164 6174   a study metadat
+000014f0: 613a 0a0a 6060 6070 7974 686f 6e0a 6d65  a:..```python.me
+00001500: 7461 6461 7461 203d 2073 7475 6479 2e6d  tadata = study.m
+00001510: 6574 616c 6973 745b 345d 0a6d 6574 6164  etalist[4].metad
+00001520: 6174 610a 6060 600a 3e60 6060 0a3e 4d65  ata.```.>```.>Me
+00001530: 7461 6461 7461 2869 643d 342c 6e61 6d65  tadata(id=4,name
+00001540: 3d22 4175 7468 6f72 2773 2063 656c 6c20  ="Author's cell 
+00001550: 7479 7065 222c 7479 7065 3d22 4361 7465  type",type="Cate
+00001560: 676f 7279 2229 0a60 6060 7079 7468 6f6e  gory").```python
+00001570: 0a6d 6574 6164 6174 612e 6665 7463 6828  .metadata.fetch(
+00001580: 290a 6d65 7461 6461 7461 2e76 616c 7565  ).metadata.value
+00001590: 730a 6060 600a 3e60 6060 0a3e 2061 7272  s.```.>```.> arr
+000015a0: 6179 285b 2743 4438 2054 2063 656c 6c73  ay(['CD8 T cells
+000015b0: 272c 2027 4465 6e64 7269 7469 6320 6365  ', 'Dendritic ce
+000015c0: 6c6c 7327 2c20 2743 4434 2054 2063 656c  lls', 'CD4 T cel
+000015d0: 6c73 272c 202e 2e2e 2c0a 3e20 2020 2020  ls', ...,.>     
+000015e0: 2020 2027 4344 3820 5420 6365 6c6c 7327     'CD8 T cells'
+000015f0: 2c20 2742 2063 656c 6c73 272c 2027 4344  , 'B cells', 'CD
+00001600: 3420 5420 6365 6c6c 7327 5d2c 2064 7479  4 T cells'], dty
+00001610: 7065 3d27 3c55 3137 2729 0a0a 2323 2320  pe='<U17')..### 
+00001620: 5175 6572 7920 6765 6e65 733a 0a0a 6060  Query genes:..``
+00001630: 6070 7974 686f 6e0a 6578 705f 6d74 7820  `python.exp_mtx 
+00001640: 3d20 7374 7564 792e 7175 6572 795f 6765  = study.query_ge
+00001650: 6e65 7328 5b27 4344 3344 272c 2027 4344  nes(['CD3D', 'CD
+00001660: 3345 275d 2c20 6269 6f66 6c65 782e 554e  3E'], bioflex.UN
+00001670: 4954 5f4c 4f47 4e4f 524d 290a 6578 705f  IT_LOGNORM).exp_
+00001680: 6d74 780a 6060 600a 3e60 6060 0a3e 203c  mtx.```.>```.> <
+00001690: 3239 3036 3578 3220 7370 6172 7365 206d  29065x2 sparse m
+000016a0: 6174 7269 7820 6f66 2074 7970 6520 273c  atrix of type '<
+000016b0: 636c 6173 7320 276e 756d 7079 2e66 6c6f  class 'numpy.flo
+000016c0: 6174 3332 273e 270a 3e20 2020 2020 7769  at32'>'.>     wi
+000016d0: 7468 2031 3534 3932 2073 746f 7265 6420  th 15492 stored 
+000016e0: 656c 656d 656e 7473 2069 6e20 436f 6d70  elements in Comp
+000016f0: 7265 7373 6564 2053 7061 7273 6520 436f  ressed Sparse Co
+00001700: 6c75 6d6e 2066 6f72 6d61 743e 0a0a 2323  lumn format>..##
+00001710: 2320 4765 7420 7374 7564 7920 6261 7263  # Get study barc
+00001720: 6f64 6573 3a0a 0a60 6060 7079 7468 6f6e  odes:..```python
+00001730: 0a73 7475 6479 2e62 6172 636f 6465 7328  .study.barcodes(
+00001740: 290a 6060 600a 3e60 6060 0a3e 205b 2747  ).```.>```.> ['G
+00001750: 534d 3235 3630 3234 395f 4141 4143 4154  SM2560249_AAACAT
+00001760: 4143 4341 4147 4354 2d31 272c 0a3e 2020  ACCAAGCT-1',.>  
+00001770: 2747 534d 3235 3630 3234 395f 4141 4143  'GSM2560249_AAAC
+00001780: 4154 4143 4343 4354 4143 2d31 272c 0a3e  ATACCCCTAC-1',.>
+00001790: 2020 2e2e 2e0a 3e20 2027 4753 4d32 3536    ....>  'GSM256
+000017a0: 3032 3439 5f41 4154 5447 5447 4154 5443  0249_AATTGTGATTC
+000017b0: 4143 542d 3127 2c0a 3e20 2027 4753 4d32  ACT-1',.>  'GSM2
+000017c0: 3536 3032 3439 5f41 4154 5447 5447 4154  560249_AATTGTGAT
+000017d0: 5454 4347 542d 3127 2c0a 3e20 202e 2e2e  TTCGT-1',.>  ...
+000017e0: 5d0a 0a23 2323 2047 6574 2073 7475 6479  ]..### Get study
+000017f0: 2066 6561 7475 7265 733a 0a0a 6060 6070   features:..```p
+00001800: 7974 686f 6e0a 7374 7564 792e 6665 6174  ython.study.feat
+00001810: 7572 6573 2829 0a60 6060 0a3e 6060 600a  ures().```.>```.
+00001820: 3e20 5b27 3553 5f52 524e 4127 2c0a 3e20  > ['5S_RRNA',.> 
+00001830: 2027 355f 3853 5f52 524e 4127 2c0a 3e20   '5_8S_RRNA',.> 
+00001840: 202e 2e2e 0a3e 2020 2741 4330 3036 3237   ....>  'AC00627
+00001850: 3327 2c0a 3e20 2027 4143 3030 3632 3737  3',.>  'AC006277
+00001860: 272c 0a3e 2020 2e2e 2e5d 0a0a 2323 2320  ',.>  ...]..### 
+00001870: 4765 7420 7374 7564 7920 6675 6c6c 206d  Get study full m
+00001880: 6174 7269 783a 0a0a 6060 6070 7974 686f  atrix:..```pytho
+00001890: 6e0a 7374 7564 792e 6d61 7472 6978 2862  n.study.matrix(b
+000018a0: 696f 666c 6578 2e55 4e49 545f 4c4f 474e  ioflex.UNIT_LOGN
+000018b0: 4f52 4d29 0a60 6060 0a3e 6060 600a 3e20  ORM).```.>```.> 
+000018c0: 3c32 3930 3635 7836 3436 3432 2073 7061  <29065x64642 spa
+000018d0: 7273 6520 6d61 7472 6978 206f 6620 7479  rse matrix of ty
+000018e0: 7065 2027 3c63 6c61 7373 2027 6e75 6d70  pe '<class 'nump
+000018f0: 792e 666c 6f61 7433 3227 3e27 0a3e 2009  y.float32'>'.> .
+00001900: 7769 7468 2031 3735 3730 3733 3920 7374  with 17570739 st
+00001910: 6f72 6564 2065 6c65 6d65 6e74 7320 696e  ored elements in
+00001920: 2043 6f6d 7072 6573 7365 6420 5370 6172   Compressed Spar
+00001930: 7365 2043 6f6c 756d 6e20 666f 726d 6174  se Column format
+00001940: 3e0a 0a23 2323 2045 7870 6f72 7420 5374  >..### Export St
+00001950: 7564 793a 0a0a 6060 6070 7974 686f 6e0a  udy:..```python.
+00001960: 7374 7564 792e 6578 706f 7274 5f73 7475  study.export_stu
+00001970: 6479 2862 696f 666c 6578 2e45 5850 4f52  dy(bioflex.EXPOR
+00001980: 545f 4835 4144 290a 6060 600a 3e60 6060  T_H5AD).```.>```
+00001990: 0a3e 7b27 646f 776e 6c6f 6164 5f6c 696e  .>{'download_lin
+000019a0: 6b27 3a20 2768 7474 7073 3a2f 2f74 616c  k': 'https://tal
+000019b0: 6b32 6461 7461 2e62 696f 7475 7269 6e67  k2data.bioturing
+000019c0: 2e63 6f6d 2f61 7069 2f65 7870 6f72 742f  .com/api/export/
+000019d0: 6131 3030 3362 6164 3364 6431 3436 6232  a1003bad3dd146b2
+000019e0: 3863 3762 6461 3931 3361 3266 6333 6630  8c7bda913a2fc3f0
+000019f0: 272c 0a3e 2027 7374 7564 795f 6861 7368  ',.> 'study_hash
+00001a00: 5f69 6427 3a20 2747 5345 3936 3538 335f  _id': 'GSE96583_
+00001a10: 6261 7463 6832 277d 0a0a 2d2d 2d2d 0a46  batch2'}..----.F
+00001a20: 6f72 2066 7572 7468 6572 2069 6e66 6f72  or further infor
+00001a30: 6d61 7469 6f6e 2070 6c65 6173 6520 6368  mation please ch
+00001a40: 6563 6b20 7468 6520 5b64 6f63 756d 656e  eck the [documen
+00001a50: 7461 7469 6f6e 5d28 6874 7470 733a 2f2f  tation](https://
+00001a60: 636f 6c61 622e 6269 6f74 7572 696e 672e  colab.bioturing.
+00001a70: 636f 6d2f 292e                           com/).
```

### Comparing `biocolabsdk-1.0.1/biocolabsdk/connector.py` & `biocolabsdk-1.0.2/biocolabsdk/connector.py`

 * *Files 27% similar despite different names*

```diff
@@ -1,164 +1,131 @@
-import os
-import pwd
-import getpass
-import json
-import grp
 import bioflex
 from bioturing_connector.connector import BBrowserXConnector
-
-
-class Identity():
-
-    def __init__(self, user: str, group: str = None):
-        self.uid = pwd.getpwnam(user).pw_uid
-        if not group:
-            self.gid = pwd.getpwnam(user).pw_gid
-        else:
-            self.gid = grp.getgrnam(group).gr_gid
-
-    def __enter__(self):
-        self.original_uid = os.getuid()
-        self.original_gid = os.getgid()
-        os.setegid(self.uid)
-        os.seteuid(self.gid)
-
-    def __exit__(self, type, value, traceback):
-        os.seteuid(self.original_uid)
-        os.setegid(self.original_gid)
-
+from .util import empty, get_settings
 
 class EConnector:
     """Create a connector object to submit/get data from BioColab
     """
 
-    def __init__(self, public_token: str = "", private_host: str = "", private_token: str = ""):
+    def __init__(self, bioflex_token: str = "", public_token: str = "", private_host: str = "", private_token: str = ""):
         """
           Args:
+            bioflex_token:
+                The API token to access BioFlex
             public_token:
-              The API token to verify authority. Generated in https://talk2data.bioturing.com
+                The API token to verify authority. Generated in https://talk2data.bioturing.com
             private_host:
-              The URL of the BBrowserX server, only supports HTTPS connection
-              Example: https://bbrowserx.bioturing.com
+                The URL of the BBrowserX server, only supports HTTPS connection
+                Example: https://bbrowserx.bioturing.com
             private_token:
-              The API token to verify authority. Generated in-app.
+                The API token to verify authority. Generated in-app.
         """
+        self.__bioflex_token = bioflex_token
         self.__public_token = public_token
         self.__private_host = private_host
         self.__private_token = private_token
 
-    def empty(self, data: str):
-        if (data is None) or (len(data) == 0):
-            return True
-        return False
-
-    def get_system_user(self):
-        try:
-            user = pwd.getpwuid(os.getuid())[0]
-        except:
-            user = os.environ.get('NB_USER', getpass.getuser())
-        return user
-
-    def get_user_home_path(self):
-        username = self.get_system_user()
-        if username is None:
-            return None
-
-        pathDir = f"/home/{username}"
-        try:
-            with Identity(username):
-                pathDir = os.path.expanduser(
-                    f"~{pwd.getpwuid(os.geteuid())[0]}")
-        except:
-            pass
-
-        return pathDir
-
-    def get_settings(self):
-        obj = {}
-        user_home_path = self.get_user_home_path()
-        if user_home_path is None:
-            return obj
-
-        bio_key_file = f"{user_home_path}/.bioapi.key"
-        if os.path.exists(bio_key_file):
-            with open(bio_key_file) as f:
-                try:
-                    obj = json.loads(f.read().strip())
-                except:
-                    pass
-                f.close()
-
-        return obj
-
     """
       Args:
         private_host:
-          The URL of the BBrowserX server, only supports HTTPS connection
-          Example: https://bbrowserx.bioturing.com
+          The URL of the BBrowserX private server, only supports HTTPS connection
+          Example: https://bbrowserx.your_domain
           If empty, BioColab will retrieve this information from your settings.
         private_token:
           The API token to verify authority. Generated in-app.
           If empty, BioColab will retrieve this information from your settings.
         ssl: SSL mode
     """
 
     def get_bbrowserx(
         self,
         private_host: str = "",
         private_token: str = "",
         ssl: bool = True
     ):
         host = private_host
-        if self.empty(host) == True:
+        if empty(host) == True:
             host = self.__private_host
 
         token = private_token
-        if self.empty(token) == True:
+        if empty(token) == True:
             token = self.__private_token
 
-        if (self.empty(host) == True) or (self.empty(token) == True):
+        if (empty(host) == True) or (empty(token) == True):
             try:
-                contents = self.get_settings()
+                contents = get_settings()
                 if "bio_bbx_private_domain" in contents:
                     host = contents["bio_bbx_private_domain"]
                 if "bio_bbx_private_key" in contents:
                     token = contents["bio_bbx_private_key"]
             except Exception as e:
                 raise e
 
-        if (self.empty(host) == True) or (self.empty(token) == True):
-            raise Exception("Empty BBrowserX setting")
+        if (empty(host) == True) or (empty(token) == True):
+            raise Exception("Empty BBrowserX setting (Host - Token)")
 
         return BBrowserXConnector(
             host=host,
             token=token,
             ssl=ssl
         )
 
     """
       Args:
         public_token:
-          The API token to verify authority. Generated in https://talk2data.bioturing.com
+          The API token to verify authority. Generated in talk2data.bioturing.com.
           If empty, BioColab will retrieve this information from your settings.
     """
 
-    def get_bioflex(
+    def get_talk2data(
         self,
         public_token: str = ""
     ):
+        host = "https://talk2data.bioturing.com/t2d_index_tool/"
         token = public_token
-        if self.empty(token) == True:
+        if empty(token) == True:
             token = self.__public_token
 
-        if (self.empty(token) == True):
+        if empty(token) == True:
+            try:
+                contents = get_settings()
+                if "bio_bbx_t2d_key" in contents:
+                    token = contents["bio_bbx_t2d_key"]
+            except Exception as e:
+                raise e
+
+        if empty(token) == True:
+            raise Exception("Empty BBrowserX setting (Public Token)")
+
+        return BBrowserXConnector(
+            host=host,
+            token=token,
+            ssl=True
+        )
+
+    """
+      Args:
+        bioflex_token:
+          The API token to verify authority. Generated in https://colab.bioturing.com
+          If empty, BioColab will retrieve this information from your settings.
+    """
+
+    def get_bioflex(
+        self,
+        bioflex_token: str = ""
+    ):
+        token = bioflex_token
+        if empty(token) == True:
+            token = self.__bioflex_token
+
+        if (empty(token) == True):
             try:
-                contents = self.get_settings()
-                if "bio_bbx_public_key" in contents:
-                    token = contents["bio_bbx_public_key"]
+                contents = get_settings()
+                if "bio_bioflex_public_key" in contents:
+                    token = contents["bio_bioflex_public_key"]
             except Exception as e:
                 raise e
 
-        if (self.empty(token) == True):
-            raise Exception("Empty BioFlex setting")
+        if (empty(token) == True):
+            raise Exception("Empty BioFlex setting (TOKEN)")
 
-        return bioflex.connect(public_token)
+        return bioflex.connect(token)
```

### Comparing `biocolabsdk-1.0.1/PKG-INFO` & `biocolabsdk-1.0.2/PKG-INFO`

 * *Files 22% similar despite different names*

```diff
@@ -1,15 +1,15 @@
 00000000: 4d65 7461 6461 7461 2d56 6572 7369 6f6e  Metadata-Version
 00000010: 3a20 322e 310a 4e61 6d65 3a20 6269 6f63  : 2.1.Name: bioc
 00000020: 6f6c 6162 7364 6b0a 5665 7273 696f 6e3a  olabsdk.Version:
-00000030: 2031 2e30 2e31 0a53 756d 6d61 7279 3a20   1.0.1.Summary: 
+00000030: 2031 2e30 2e32 0a53 756d 6d61 7279 3a20   1.0.2.Summary: 
 00000040: 4120 7365 7420 6f66 2070 7974 686f 6e20  A set of python 
 00000050: 6d6f 6475 6c65 7320 666f 7220 6163 6365  modules for acce
 00000060: 7373 696e 6720 4269 6f54 7572 696e 6720  ssing BioTuring 
-00000070: 6563 6f73 7973 7465 6d20 6f6e 2042 696f  ecosystem on Bio
+00000070: 4563 6f73 7973 7465 6d20 6f6e 2042 696f  Ecosystem on Bio
 00000080: 436f 6c61 6220 7072 6976 6174 6520 7365  Colab private se
 00000090: 7276 6572 0a41 7574 686f 723a 2042 696f  rver.Author: Bio
 000000a0: 5475 7269 6e67 0a41 7574 686f 722d 656d  Turing.Author-em
 000000b0: 6169 6c3a 2073 7570 706f 7274 4062 696f  ail: support@bio
 000000c0: 7475 7269 6e67 2e63 6f6d 0a52 6571 7569  turing.com.Requi
 000000d0: 7265 732d 5079 7468 6f6e 3a20 3e3d 332e  res-Python: >=3.
 000000e0: 372c 3c34 2e30 0a43 6c61 7373 6966 6965  7,<4.0.Classifie
@@ -35,485 +35,431 @@
 00000220: 3a20 6269 6f66 6c65 7820 283e 3d31 2e31  : bioflex (>=1.1
 00000230: 2c3c 322e 3029 0a52 6571 7569 7265 732d  ,<2.0).Requires-
 00000240: 4469 7374 3a20 6269 6f74 7572 696e 672d  Dist: bioturing-
 00000250: 636f 6e6e 6563 746f 7220 283e 3d31 2e31  connector (>=1.1
 00000260: 2c3c 322e 3029 0a44 6573 6372 6970 7469  ,<2.0).Descripti
 00000270: 6f6e 2d43 6f6e 7465 6e74 2d54 7970 653a  on-Content-Type:
 00000280: 2074 6578 742f 6d61 726b 646f 776e 0a0a   text/markdown..
-00000290: 2323 2031 2e20 5573 6167 6520 6269 6f63  ## 1. Usage bioc
-000002a0: 6f6c 6162 7364 6b3a 0a2a 2a54 6865 2070  olabsdk:.**The p
-000002b0: 6163 6b61 6765 206f 6e6c 7920 616c 6c6f  ackage only allo
-000002c0: 7773 2064 6174 6120 7375 626d 6973 7369  ws data submissi
-000002d0: 6f6e 2076 6961 2042 696f 436f 6c61 6220  on via BioColab 
-000002e0: 7072 6976 6174 6520 7365 7276 6572 2e20  private server. 
-000002f0: 506c 6561 7365 2063 6f6e 6669 6775 7265  Please configure
-00000300: 2079 6f75 7220 746f 6b65 6e73 2069 6e20   your tokens in 
-00000310: 7468 6520 6055 7365 7220 5365 7474 696e  the `User Settin
-00000320: 6773 6020 7061 6765 2e2a 2a0a 2323 2320  gs` page.**.### 
-00000330: 312e 312e 2054 6573 7420 7468 6520 636f  1.1. Test the co
-00000340: 6e6e 6563 7469 6f6e 3a0a 6060 600a 2320  nnection:.```.# 
-00000350: 6578 616d 706c 652e 7079 0a0a 6672 6f6d  example.py..from
-00000360: 2062 696f 636f 6c61 6273 646b 2e63 6f6e   biocolabsdk.con
-00000370: 6e65 6374 6f72 2069 6d70 6f72 7420 4543  nector import EC
-00000380: 6f6e 6e65 6374 6f72 0a0a 636f 6e6e 6563  onnector..connec
-00000390: 746f 7220 3d20 4543 6f6e 6e65 6374 6f72  tor = EConnector
-000003a0: 280a 2020 686f 7374 3d22 6874 7470 733a  (.  host="https:
-000003b0: 2f2f 796f 7572 636f 6d70 616e 792f 7432  //yourcompany/t2
-000003c0: 645f 696e 6465 785f 746f 6f6c 2f2c 0a20  d_index_tool/,. 
-000003d0: 2074 6f6b 656e 3d22 3c69 6e70 7574 2079   token="<input y
-000003e0: 6f75 7220 746f 6b65 6e20 6865 7265 3e22  our token here>"
-000003f0: 0a29 0a0a 636f 6e6e 6563 746f 722e 6765  .)..connector.ge
-00000400: 745f 6262 726f 7773 6572 7828 292e 7465  t_bbrowserx().te
-00000410: 7374 5f63 6f6e 6e65 6374 696f 6e28 290a  st_connection().
-00000420: 6060 600a 0a45 7861 6d70 6c65 206f 7574  ```..Example out
-00000430: 7075 743a 0a0a 6060 600a 436f 6e6e 6563  put:..```.Connec
-00000440: 7469 6e67 2074 6f20 686f 7374 2061 7420  ting to host at 
-00000450: 6874 7470 733a 2f2f 796f 7572 636f 6d70  https://yourcomp
-00000460: 616e 792f 7432 645f 696e 6465 785f 746f  any/t2d_index_to
-00000470: 6f6c 2f61 7069 2f76 312f 7465 7374 5f63  ol/api/v1/test_c
-00000480: 6f6e 6e65 6374 696f 6e0a 436f 6e6e 6563  onnection.Connec
-00000490: 7469 6f6e 2073 7563 6365 7373 6675 6c0a  tion successful.
-000004a0: 6060 600a 0a23 2323 2031 2e32 2e20 4765  ```..### 1.2. Ge
-000004b0: 7420 7573 6572 2067 726f 7570 7320 6176  t user groups av
-000004c0: 6169 6c61 626c 6520 666f 7220 796f 7572  ailable for your
-000004d0: 2074 6f6b 656e 3a0a 6060 600a 2320 6578   token:.```.# ex
-000004e0: 616d 706c 652e 7079 0a0a 6672 6f6d 2062  ample.py..from b
-000004f0: 696f 636f 6c61 6273 646b 2e63 6f6e 6e65  iocolabsdk.conne
-00000500: 6374 6f72 2069 6d70 6f72 7420 4543 6f6e  ctor import ECon
-00000510: 6e65 6374 6f72 0a0a 636f 6e6e 6563 746f  nector..connecto
-00000520: 7220 3d20 4543 6f6e 6e65 6374 6f72 280a  r = EConnector(.
-00000530: 2020 686f 7374 3d22 6874 7470 733a 2f2f    host="https://
-00000540: 796f 7572 636f 6d70 616e 792f 7432 645f  yourcompany/t2d_
-00000550: 696e 6465 785f 746f 6f6c 2f2c 0a20 2074  index_tool/,.  t
-00000560: 6f6b 656e 3d22 3c69 6e70 7574 2079 6f75  oken="<input you
-00000570: 7220 746f 6b65 6e20 6865 7265 3e22 0a29  r token here>".)
-00000580: 0a0a 7573 6572 5f67 726f 7570 7320 3d20  ..user_groups = 
-00000590: 636f 6e6e 6563 746f 722e 6765 745f 6262  connector.get_bb
-000005a0: 726f 7773 6572 7828 292e 6765 745f 7573  rowserx().get_us
-000005b0: 6572 5f67 726f 7570 7328 290a 7072 696e  er_groups().prin
-000005c0: 7428 7573 6572 5f67 726f 7570 7329 0a60  t(user_groups).`
-000005d0: 6060 0a0a 4578 616d 706c 6520 6f75 7470  ``..Example outp
-000005e0: 7574 3a0a 0a60 6060 0a5b 7b27 6964 273a  ut:..```.[{'id':
-000005f0: 2027 616c 6c5f 6d65 6d62 6572 7327 2c20   'all_members', 
-00000600: 276e 616d 6527 3a20 2741 6c6c 206d 656d  'name': 'All mem
-00000610: 6265 7273 277d 2c20 7b27 6964 273a 2027  bers'}, {'id': '
-00000620: 7065 7273 6f6e 616c 272c 2027 6e61 6d65  personal', 'name
-00000630: 273a 2027 5065 7273 6f6e 616c 2077 6f72  ': 'Personal wor
-00000640: 6b73 7061 6365 277d 5d0a 6060 600a 0a23  kspace'}].```..#
-00000650: 2323 2031 2e33 2e20 5375 626d 6974 2068  ## 1.3. Submit h
-00000660: 3561 6420 2873 6361 6e70 7920 6f62 6a65  5ad (scanpy obje
-00000670: 6374 293a 0a60 6060 0a23 2065 7861 6d70  ct):.```.# examp
-00000680: 6c65 2e70 790a 6672 6f6d 2062 696f 636f  le.py.from bioco
-00000690: 6c61 6273 646b 2e63 6f6e 6e65 6374 6f72  labsdk.connector
-000006a0: 2069 6d70 6f72 7420 4543 6f6e 6e65 6374   import EConnect
-000006b0: 6f72 0a66 726f 6d20 6269 6f74 7572 696e  or.from bioturin
-000006c0: 675f 636f 6e6e 6563 746f 722e 7479 7069  g_connector.typi
-000006d0: 6e67 2069 6d70 6f72 7420 496e 7075 744d  ng import InputM
-000006e0: 6174 7269 7854 7970 650a 6672 6f6d 2062  atrixType.from b
-000006f0: 696f 7475 7269 6e67 5f63 6f6e 6e65 6374  ioturing_connect
-00000700: 6f72 2e74 7970 696e 6720 696d 706f 7274  or.typing import
-00000710: 2053 7065 6369 6573 0a0a 636f 6e6e 6563   Species..connec
-00000720: 746f 7220 3d20 4543 6f6e 6e65 6374 6f72  tor = EConnector
-00000730: 280a 2020 686f 7374 3d22 6874 7470 733a  (.  host="https:
-00000740: 2f2f 796f 7572 636f 6d70 616e 792f 7432  //yourcompany/t2
-00000750: 645f 696e 6465 785f 746f 6f6c 2f2c 0a20  d_index_tool/,. 
-00000760: 2074 6f6b 656e 3d22 3c69 6e70 7574 2079   token="<input y
-00000770: 6f75 7220 746f 6b65 6e20 6865 7265 3e22  our token here>"
-00000780: 0a29 0a0a 2320 4361 6c6c 2074 6869 7320  .)..# Call this 
-00000790: 6675 6e63 7469 6f6e 2066 6972 7374 2074  function first t
-000007a0: 6f20 6765 7420 6176 6169 6c61 626c 6520  o get available 
-000007b0: 6772 6f75 7073 2061 6e64 2074 6865 6972  groups and their
-000007c0: 2069 642e 0a75 7365 725f 6772 6f75 7073   id..user_groups
-000007d0: 203d 2063 6f6e 6e65 6374 6f72 2e67 6574   = connector.get
-000007e0: 5f62 6272 6f77 7365 7278 2829 2e67 6574  _bbrowserx().get
-000007f0: 5f75 7365 725f 6772 6f75 7073 2829 0a23  _user_groups().#
-00000800: 2045 7861 6d70 6c65 3a20 7573 6572 5f67   Example: user_g
-00000810: 726f 7570 7320 6973 206e 6f77 205b 7b27  roups is now [{'
-00000820: 6964 273a 2027 616c 6c5f 6d65 6d62 6572  id': 'all_member
-00000830: 7327 2c20 276e 616d 6527 3a20 2741 6c6c  s', 'name': 'All
-00000840: 206d 656d 6265 7273 277d 2c20 7b27 6964   members'}, {'id
-00000850: 273a 2027 7065 7273 6f6e 616c 272c 2027  ': 'personal', '
-00000860: 6e61 6d65 273a 2027 5065 7273 6f6e 616c  name': 'Personal
-00000870: 2077 6f72 6b73 7061 6365 277d 5d0a 0a0a   workspace'}]...
-00000880: 2320 5375 626d 6974 7469 6e67 2074 6865  # Submitting the
-00000890: 2073 6361 6e70 7920 6f62 6a65 6374 3a0a   scanpy object:.
-000008a0: 636f 6e6e 6563 746f 722e 7375 626d 6974  connector.submit
-000008b0: 5f68 3561 6428 0a20 2067 726f 7570 5f69  _h5ad(.  group_i
-000008c0: 643d 2770 6572 736f 6e61 6c27 2c0a 2020  d='personal',.  
-000008d0: 7374 7564 795f 7333 5f6b 6579 733d 5b27  study_s3_keys=['
-000008e0: 4753 4531 3238 3232 332e 6835 6164 275d  GSE128223.h5ad']
-000008f0: 2c0a 2020 7374 7564 795f 6964 3d27 4753  ,.  study_id='GS
-00000900: 4531 3238 3232 3327 2c0a 2020 6e61 6d65  E128223',.  name
-00000910: 3d27 5468 6973 2069 7320 6d79 2066 6972  ='This is my fir
-00000920: 7374 2073 7475 6479 272c 0a20 2061 7574  st study',.  aut
-00000930: 686f 7273 3d5b 2748 7579 204e 6775 7965  hors=['Huy Nguye
-00000940: 6e27 5d2c 0a20 2073 7065 6369 6573 3d53  n'],.  species=S
-00000950: 7065 6369 6573 2e48 554d 414e 2e76 616c  pecies.HUMAN.val
-00000960: 7565 2c0a 2020 696e 7075 745f 6d61 7472  ue,.  input_matr
-00000970: 6978 5f74 7970 653d 496e 7075 744d 6174  ix_type=InputMat
-00000980: 7269 7854 7970 652e 5241 572e 7661 6c75  rixType.RAW.valu
-00000990: 650a 290a 0a23 2045 7861 6d70 6c65 206f  e.)..# Example o
-000009a0: 7574 7075 743a 0a3e 205b 3230 3232 2d31  utput:.> [2022-1
-000009b0: 302d 3130 2030 313a 3033 5d20 5761 6974  0-10 01:03] Wait
-000009c0: 696e 6720 696e 2071 7565 7565 0a3e 205b  ing in queue.> [
-000009d0: 3230 3232 2d31 302d 3130 2030 313a 3033  2022-10-10 01:03
-000009e0: 5d20 446f 776e 6c6f 6164 696e 6720 4753  ] Downloading GS
-000009f0: 4531 3238 3232 332e 6835 6164 2066 726f  E128223.h5ad fro
-00000a00: 6d20 7333 3a20 3236 322e 3120 4b42 202f  m s3: 262.1 KB /
-00000a10: 2034 3332 2e38 204d 420a 3e20 5b32 3032   432.8 MB.> [202
-00000a20: 322d 3130 2d31 3020 3031 3a30 335d 2046  2-10-10 01:03] F
-00000a30: 696c 6520 646f 776e 6c6f 6164 6564 0a3e  ile downloaded.>
-00000a40: 205b 3230 3232 2d31 302d 3130 2030 313a   [2022-10-10 01:
-00000a50: 3033 5d20 5265 6164 696e 6720 6261 7463  03] Reading batc
-00000a60: 683a 2047 5345 3132 3832 3233 2e68 3561  h: GSE128223.h5a
-00000a70: 640a 3e20 5b32 3032 322d 3130 2d31 3020  d.> [2022-10-10 
-00000a80: 3031 3a30 335d 2050 7265 7072 6f63 6573  01:03] Preproces
-00000a90: 7369 6e67 2065 7870 7265 7373 696f 6e20  sing expression 
-00000aa0: 6d61 7472 6978 3a20 3139 3132 3120 6365  matrix: 19121 ce
-00000ab0: 6c6c 7320 7820 3633 3831 3320 6765 6e65  lls x 63813 gene
-00000ac0: 730a 3e20 5b32 3032 322d 3130 2d31 3020  s.> [2022-10-10 
-00000ad0: 3031 3a30 335d 2046 696c 7465 7265 643a  01:03] Filtered:
-00000ae0: 2031 3931 3231 2063 656c 6c73 2072 656d   19121 cells rem
-00000af0: 6169 6e0a 3e20 5b32 3032 322d 3130 2d31  ain.> [2022-10-1
-00000b00: 3020 3031 3a30 335d 2053 7461 7274 2070  0 01:03] Start p
-00000b10: 726f 6365 7373 696e 6720 7374 7564 790a  rocessing study.
-00000b20: 3e20 5b32 3032 322d 3130 2d31 3020 3031  > [2022-10-10 01
-00000b30: 3a30 335d 204e 6f72 6d61 6c69 7a69 6e67  :03] Normalizing
-00000b40: 2065 7870 7265 7373 696f 6e20 6d61 7472   expression matr
-00000b50: 6978 0a3e 205b 3230 3232 2d31 302d 3130  ix.> [2022-10-10
-00000b60: 2030 313a 3033 5d20 5275 6e6e 696e 6720   01:03] Running 
-00000b70: 5043 410a 3e20 5b32 3032 322d 3130 2d31  PCA.> [2022-10-1
-00000b80: 3020 3031 3a30 335d 2052 756e 6e69 6e67  0 01:03] Running
-00000b90: 206b 4e4e 0a3e 205b 3230 3232 2d31 302d   kNN.> [2022-10-
-00000ba0: 3130 2030 313a 3033 5d20 5275 6e6e 696e  10 01:03] Runnin
-00000bb0: 6720 7370 6563 7472 616c 2065 6d62 6564  g spectral embed
-00000bc0: 6469 6e67 0a3e 205b 3230 3232 2d31 302d  ding.> [2022-10-
-00000bd0: 3130 2030 313a 3033 5d20 5275 6e6e 696e  10 01:03] Runnin
-00000be0: 6720 7665 6e69 6365 2062 696e 6172 697a  g venice binariz
-00000bf0: 6572 0a3e 205b 3230 3232 2d31 302d 3130  er.> [2022-10-10
-00000c00: 2030 313a 3034 5d20 5275 6e6e 696e 6720   01:04] Running 
-00000c10: 742d 534e 450a 3e20 5b32 3032 322d 3130  t-SNE.> [2022-10
-00000c20: 2d31 3020 3031 3a30 345d 2053 7475 6479  -10 01:04] Study
-00000c30: 2077 6173 2073 7563 6365 7373 6675 6c6c   was successfull
-00000c40: 7920 7375 626d 6974 7465 640a 3e20 5b32  y submitted.> [2
-00000c50: 3032 322d 3130 2d31 3020 3031 3a30 345d  022-10-10 01:04]
-00000c60: 2044 4f4e 4520 2121 210a 3e20 5374 7564   DONE !!!.> Stud
-00000c70: 7920 7375 626d 6974 7465 6420 7375 6363  y submitted succ
-00000c80: 6573 7366 756c 6c79 210a 6060 600a 4176  essfully!.```.Av
-00000c90: 6169 6c61 626c 6520 7061 7261 6d65 7465  ailable paramete
-00000ca0: 7273 2066 6f72 2060 7375 626d 6974 5f68  rs for `submit_h
-00000cb0: 3561 6460 2066 756e 6374 696f 6e3a 0a60  5ad` function:.`
-00000cc0: 6060 0a67 726f 7570 5f69 643a 2073 7472  ``.group_id: str
-00000cd0: 0a20 2049 4420 6f66 2074 6865 2067 726f  .  ID of the gro
-00000ce0: 7570 2074 6f20 7375 626d 6974 2074 6865  up to submit the
-00000cf0: 2064 6174 6120 746f 2e0a 0a73 7475 6479   data to...study
-00000d00: 5f73 335f 6b65 7973 3a20 4c69 7374 5b73  _s3_keys: List[s
-00000d10: 7472 5d0a 2020 4c69 7374 206f 6620 7468  tr].  List of th
-00000d20: 6520 7333 206b 6579 206f 6620 7468 6520  e s3 key of the 
-00000d30: 7374 7564 6965 732e 0a0a 7374 7564 795f  studies...study_
-00000d40: 6964 3a20 7374 722c 2064 6566 6175 6c74  id: str, default
-00000d50: 3d4e 6f6e 650a 2020 5374 7564 7920 4944  =None.  Study ID
-00000d60: 2c20 6966 206e 6f20 7661 6c75 6520 6973  , if no value is
-00000d70: 2073 7065 6369 6669 6564 2c20 7573 6520   specified, use 
-00000d80: 6120 7261 6e64 6f6d 2075 7569 6476 3420  a random uuidv4 
-00000d90: 7374 7269 6e67 0a0a 6e61 6d65 3a20 7374  string..name: st
-00000da0: 722c 2064 6566 6175 6c74 3d27 546f 2062  r, default='To b
-00000db0: 6520 6465 7461 696c 6564 270a 2020 4e61  e detailed'.  Na
-00000dc0: 6d65 206f 6620 7468 6520 7374 7564 792e  me of the study.
-00000dd0: 0a0a 6175 7468 6f72 733a 204c 6973 745b  ..authors: List[
-00000de0: 7374 725d 2c20 6465 6661 756c 743d 5b5d  str], default=[]
-00000df0: 0a20 2041 7574 686f 7273 206f 6620 7468  .  Authors of th
-00000e00: 6520 7374 7564 792e 0a0a 6162 7374 7261  e study...abstra
-00000e10: 6374 3a20 7374 722c 2064 6566 6175 6c74  ct: str, default
-00000e20: 3d27 270a 2020 4162 7374 7261 6374 206f  =''.  Abstract o
-00000e30: 6620 7468 6520 7374 7564 792e 0a0a 7370  f the study...sp
-00000e40: 6563 6965 733a 2073 7472 2c20 6465 6661  ecies: str, defa
-00000e50: 756c 743d 2768 756d 616e 270a 2020 5370  ult='human'.  Sp
-00000e60: 6563 6965 7320 6f66 2074 6865 2073 7475  ecies of the stu
-00000e70: 6479 2e20 4361 6e20 6265 3a20 2a2a 6269  dy. Can be: **bi
-00000e80: 6f74 7572 696e 675f 636f 6e6e 6563 746f  oturing_connecto
-00000e90: 722e 7479 7069 6e67 2e53 7065 6369 6573  r.typing.Species
-00000ea0: 2e48 554d 414e 2e76 616c 7565 2a2a 206f  .HUMAN.value** o
-00000eb0: 7220 2a2a 6269 6f74 7572 696e 675f 636f  r **bioturing_co
-00000ec0: 6e6e 6563 746f 722e 7479 7069 6e67 2e53  nnector.typing.S
-00000ed0: 7065 6369 6573 2e4d 4f55 5345 2e76 616c  pecies.MOUSE.val
-00000ee0: 7565 2a2a 206f 7220 2a2a 6269 6f74 7572  ue** or **biotur
-00000ef0: 696e 675f 636f 6e6e 6563 746f 722e 7479  ing_connector.ty
-00000f00: 7069 6e67 2e53 7065 6369 6573 2e4e 4f4e  ping.Species.NON
-00000f10: 5f48 554d 414e 5f50 5249 4d41 5445 2e76  _HUMAN_PRIMATE.v
-00000f20: 616c 7565 2a2a 0a0a 696e 7075 745f 6d61  alue**..input_ma
-00000f30: 7472 6978 5f74 7970 653a 2073 7472 2c20  trix_type: str, 
-00000f40: 6465 6661 756c 743d 2772 6177 270a 2020  default='raw'.  
-00000f50: 4966 2074 6865 2076 616c 7565 206f 6620  If the value of 
-00000f60: 7468 6973 2069 6e70 7574 2069 7320 2a2a  this input is **
-00000f70: 6269 6f74 7572 696e 675f 636f 6e6e 6563  bioturing_connec
-00000f80: 746f 722e 7479 7069 6e67 2e49 6e70 7574  tor.typing.Input
-00000f90: 4d61 7472 6978 5479 7065 2e4e 4f52 4d41  MatrixType.NORMA
-00000fa0: 4c49 5a45 442e 7661 6c75 652a 2a2c 0a20  LIZED.value**,. 
-00000fb0: 2074 6865 6e20 7468 6520 736f 6674 7761   then the softwa
-00000fc0: 7265 2077 696c 6c0a 2020 7573 6520 736c  re will.  use sl
-00000fd0: 6f74 2027 5827 2066 726f 6d20 7468 6520  ot 'X' from the 
-00000fe0: 7363 616e 7079 206f 626a 6563 7420 616e  scanpy object an
-00000ff0: 6420 646f 6573 206e 6f74 2061 7070 6c79  d does not apply
-00001000: 206e 6f72 6d61 6c69 7a61 7469 6f6e 2e0a   normalization..
-00001010: 2020 4966 2074 6865 2076 616c 7565 206f    If the value o
-00001020: 6620 7468 6973 2069 6e70 7574 2069 7320  f this input is 
-00001030: 2a2a 6269 6f74 7572 696e 675f 636f 6e6e  **bioturing_conn
-00001040: 6563 746f 722e 7479 7069 6e67 2e49 6e70  ector.typing.Inp
-00001050: 7574 4d61 7472 6978 5479 7065 2e52 4157  utMatrixType.RAW
-00001060: 2e76 616c 7565 2a2a 2c74 6865 6e20 7468  .value**,then th
-00001070: 6520 736f 6674 7761 7265 2077 696c 6c0a  e software will.
-00001080: 2020 7573 6520 736c 6f74 2027 7261 772e    use slot 'raw.
-00001090: 5827 2066 726f 6d20 7468 6573 6361 6e70  X' from thescanp
-000010a0: 7920 6f62 6a65 6374 2061 6e64 2061 7070  y object and app
-000010b0: 6c79 206c 6f67 2d6e 6f72 6d61 6c69 7a61  ly log-normaliza
-000010c0: 7469 6f6e 2e0a 0a6d 696e 5f63 6f75 6e74  tion...min_count
-000010d0: 733a 2069 6e74 2c20 6465 6661 756c 743d  s: int, default=
-000010e0: 4e6f 6e65 0a20 204d 696e 696d 756d 206e  None.  Minimum n
-000010f0: 756d 6265 7220 6f66 2063 6f75 6e74 7320  umber of counts 
-00001100: 7265 7175 6972 6564 0a20 2066 6f72 2061  required.  for a
-00001110: 2063 656c 6c20 746f 2070 6173 7320 6669   cell to pass fi
-00001120: 6c74 6572 696e 672e 0a0a 6d69 6e5f 6765  ltering...min_ge
-00001130: 6e65 733a 2069 6e74 2c20 6465 6661 756c  nes: int, defaul
-00001140: 743d 4e6f 6e65 0a20 204d 696e 696d 756d  t=None.  Minimum
-00001150: 206e 756d 6265 7220 6f66 2067 656e 6573   number of genes
-00001160: 2065 7870 7265 7373 6564 2072 6571 7569   expressed requi
-00001170: 7265 640a 2020 666f 7220 6120 6365 6c6c  red.  for a cell
-00001180: 2074 6f20 7061 7373 2066 696c 7465 7269   to pass filteri
-00001190: 6e67 2e0a 0a6d 6178 5f63 6f75 6e74 733a  ng...max_counts:
-000011a0: 2069 6e74 2c20 6465 6661 756c 743d 4e6f   int, default=No
-000011b0: 6e65 0a20 204d 6178 696d 756d 206e 756d  ne.  Maximum num
-000011c0: 6265 7220 6f66 2063 6f75 6e74 7320 7265  ber of counts re
-000011d0: 7175 6972 6564 0a20 2066 6f72 2061 2063  quired.  for a c
-000011e0: 656c 6c20 746f 2070 6173 7320 6669 6c74  ell to pass filt
-000011f0: 6572 696e 672e 0a0a 6d61 785f 6765 6e65  ering...max_gene
-00001200: 733a 2069 6e74 2c20 6465 6661 756c 743d  s: int, default=
-00001210: 4e6f 6e65 0a20 204d 6178 696d 756d 206e  None.  Maximum n
-00001220: 756d 6265 7220 6f66 2067 656e 6573 2065  umber of genes e
-00001230: 7870 7265 7373 6564 2072 6571 7569 7265  xpressed require
-00001240: 640a 2020 666f 7220 6120 6365 6c6c 2074  d.  for a cell t
-00001250: 6f20 7061 7373 2066 696c 7465 7269 6e67  o pass filtering
-00001260: 2e0a 0a6d 745f 7065 7263 656e 7461 6765  ...mt_percentage
-00001270: 3a20 556e 696f 6e5b 696e 742c 2066 6c6f  : Union[int, flo
-00001280: 6174 5d2c 2064 6566 6175 6c74 3d4e 6f6e  at], default=Non
-00001290: 650a 2020 4d61 7869 6d75 6d20 6e75 6d62  e.  Maximum numb
-000012a0: 6572 206f 6620 6d69 746f 6368 6f6e 6472  er of mitochondr
-000012b0: 6961 2067 656e 6573 2070 6572 6365 6e74  ia genes percent
-000012c0: 6167 650a 2020 7265 7175 6972 6564 2066  age.  required f
-000012d0: 6f72 2061 2063 656c 6c20 746f 2070 6173  or a cell to pas
-000012e0: 7320 6669 6c74 6572 696e 672e 2052 616e  s filtering. Ran
-000012f0: 6769 6e67 2066 726f 6d20 3020 746f 2031  ging from 0 to 1
-00001300: 3030 0a60 6060 0a0a 2323 2032 2e20 5573  00.```..## 2. Us
-00001310: 6167 6520 6269 6f66 6c65 783a 0a0a 2323  age bioflex:..##
-00001320: 2320 4372 6561 7465 2061 2063 6f6e 6e65  # Create a conne
-00001330: 6374 696f 6e20 7573 696e 6720 6163 6365  ction using acce
-00001340: 7373 2074 6f6b 656e 3a0a 0a60 6060 7079  ss token:..```py
-00001350: 7468 6f6e 0a66 726f 6d20 6269 6f63 6f6c  thon.from biocol
-00001360: 6162 7364 6b2e 636f 6e6e 6563 746f 7220  absdk.connector 
-00001370: 696d 706f 7274 2045 436f 6e6e 6563 746f  import EConnecto
-00001380: 720a 0a63 6f6e 6e65 6374 6f72 203d 2045  r..connector = E
-00001390: 436f 6e6e 6563 746f 7228 0a20 2070 7562  Connector(.  pub
-000013a0: 6c69 635f 746f 6b65 6e3d 223c 696e 7075  lic_token="<inpu
-000013b0: 7420 796f 7572 2074 6f6b 656e 2068 6572  t your token her
-000013c0: 653e 220a 290a 6060 600a 0a23 2323 204c  e>".).```..### L
-000013d0: 6973 7420 6176 6169 6c61 626c 6520 6461  ist available da
-000013e0: 7461 6261 7365 733a 0a0a 6060 6070 7974  tabases:..```pyt
-000013f0: 686f 6e0a 6461 7461 6261 7365 7320 3d20  hon.databases = 
-00001400: 636f 6e6e 6563 7469 6f6e 2e67 6574 5f62  connection.get_b
-00001410: 696f 666c 6578 2829 2e64 6174 6162 6173  ioflex().databas
-00001420: 6573 2829 0a60 6060 0a3e 6060 600a 3e20  es().```.>```.> 
-00001430: 5b44 6174 6142 6173 6528 6964 3d22 3530  [DataBase(id="50
-00001440: 3130 6337 6435 3733 6165 3466 6632 6239  10c7d573ae4ff2b9
-00001450: 3639 3134 3232 6239 3961 6132 6364 222c  691422b99aa2cd",
-00001460: 0a3e 2020 2020 2020 2020 2020 206e 616d  .>           nam
-00001470: 653d 2242 696f 5475 7269 6e67 2064 6174  e="BioTuring dat
-00001480: 6162 6173 6522 2c73 7065 6369 6573 3d22  abase",species="
-00001490: 6875 6d61 6e22 2c76 6572 7369 6f6e 3d31  human",version=1
-000014a0: 292c 0a3e 2044 6174 6142 6173 6528 6964  ),.> DataBase(id
-000014b0: 3d22 3530 3130 6337 6435 3733 6165 3466  ="5010c7d573ae4f
-000014c0: 6632 6239 3639 3134 3232 6239 3961 6132  f2b9691422b99aa2
-000014d0: 6364 222c 0a3e 2020 2020 2020 2020 2020  cd",.>          
-000014e0: 206e 616d 653d 2242 696f 5475 7269 6e67   name="BioTuring
-000014f0: 2064 6174 6162 6173 6522 2c73 7065 6369   database",speci
-00001500: 6573 3d22 6875 6d61 6e22 2c76 6572 7369  es="human",versi
-00001510: 6f6e 3d32 292c 0a3e 2044 6174 6142 6173  on=2),.> DataBas
-00001520: 6528 6964 3d22 3530 3130 6337 6435 3733  e(id="5010c7d573
-00001530: 6165 3466 6632 6239 3639 3134 3232 6239  ae4ff2b9691422b9
-00001540: 3961 6132 6364 222c 0a3e 2020 2020 2020  9aa2cd",.>      
-00001550: 2020 2020 206e 616d 653d 2242 696f 5475       name="BioTu
-00001560: 7269 6e67 2064 6174 6162 6173 6522 2c73  ring database",s
-00001570: 7065 6369 6573 3d22 6875 6d61 6e22 2c76  pecies="human",v
-00001580: 6572 7369 6f6e 3d33 295d 0a0a 2323 2320  ersion=3)]..### 
-00001590: 4765 7420 6461 7461 6261 7365 2063 656c  Get database cel
-000015a0: 6c20 7479 7065 7320 6765 6e65 2065 7870  l types gene exp
-000015b0: 7265 7373 696f 6e20 7375 6d6d 6172 790a  ression summary.
-000015c0: 0a60 6060 7079 7468 6f6e 0a64 6174 6162  .```python.datab
-000015d0: 6173 6520 3d20 6461 7461 6261 7365 735b  ase = databases[
-000015e0: 325d 0a64 6174 6162 6173 652e 6765 745f  2].database.get_
-000015f0: 6365 6c6c 7479 7065 735f 6578 7072 6573  celltypes_expres
-00001600: 7369 6f6e 5f73 756d 6d61 7279 285b 2743  sion_summary(['C
-00001610: 4433 4427 2c20 2743 4433 4527 5d29 0a60  D3D', 'CD3E']).`
-00001620: 6060 0a3e 6060 600a 3e20 7b27 4344 3344  ``.>```.> {'CD3D
-00001630: 273a 205b 5375 6d6d 6172 7928 6e61 6d65  ': [Summary(name
-00001640: 3d22 4220 6365 6c6c 222c 7375 6d3d 3730  ="B cell",sum=70
-00001650: 3731 3038 3837 342e 302c 6d65 616e 3d34  7108874.0,mean=4
-00001660: 3139 322e 3730 3936 2c72 6174 653d 302e  192.7096,rate=0.
-00001670: 3033 352c 636f 756e 743d 3136 3836 3532  035,count=168652
-00001680: 2e30 2c74 6f74 616c 3d34 3831 3239 3637  .0,total=4812967
-00001690: 292c 0a3e 2009 5375 6d6d 6172 7928 6e61  ),.> .Summary(na
-000016a0: 6d65 3d22 4344 342d 706f 7369 7469 7665  me="CD4-positive
-000016b0: 2c20 616c 7068 612d 6265 7461 2054 2063  , alpha-beta T c
-000016c0: 656c 6c22 2c73 756d 3d39 3438 3939 3837  ell",sum=9489987
-000016d0: 3434 322e 302c 6d65 616e 3d34 3635 372e  442.0,mean=4657.
-000016e0: 3536 3139 2c72 6174 653d 302e 3532 3833  5619,rate=0.5283
-000016f0: 2c63 6f75 6e74 3d32 3033 3735 3434 2e30  ,count=2037544.0
-00001700: 2c74 6f74 616c 3d33 3835 3635 3930 292c  ,total=3856590),
-00001710: 0a3e 2009 2e2e 2e0a 3e20 0953 756d 6d61  .> .....> .Summa
-00001720: 7279 286e 616d 653d 2263 6f72 6e65 616c  ry(name="corneal
-00001730: 2070 726f 6765 6e69 746f 7222 2c73 756d   progenitor",sum
-00001740: 3d30 2e30 2c6d 6561 6e3d 302e 302c 7261  =0.0,mean=0.0,ra
-00001750: 7465 3d30 2e30 2c63 6f75 6e74 3d30 2e30  te=0.0,count=0.0
-00001760: 2c74 6f74 616c 3d33 3937 3329 2c0a 3e20  ,total=3973),.> 
-00001770: 0953 756d 6d61 7279 286e 616d 653d 226e  .Summary(name="n
-00001780: 7563 6c65 7573 2070 756c 706f 7375 7320  ucleus pulposus 
-00001790: 7072 6f67 656e 6974 6f72 2063 656c 6c22  progenitor cell"
-000017a0: 2c73 756d 3d30 2e30 2c6d 6561 6e3d 302e  ,sum=0.0,mean=0.
-000017b0: 302c 7261 7465 3d30 2e30 2c63 6f75 6e74  0,rate=0.0,count
-000017c0: 3d30 2e30 2c74 6f74 616c 3d32 3331 3029  =0.0,total=2310)
-000017d0: 5d7d 0a0a 0a23 2323 2043 7265 6174 6520  ]}...### Create 
-000017e0: 7374 7564 7920 696e 7374 616e 6365 2c20  study instance, 
-000017f0: 7573 696e 6720 7374 7564 7920 6861 7368  using study hash
-00001800: 2049 4420 6672 6f6d 205b 4269 6f54 7572   ID from [BioTur
-00001810: 696e 6720 7374 7564 6965 735d 2868 7474  ing studies](htt
-00001820: 7073 3a2f 2f74 616c 6b32 6461 7461 2e62  ps://talk2data.b
-00001830: 696f 7475 7269 6e67 2e63 6f6d 2f73 7475  ioturing.com/stu
-00001840: 6469 6573 2f29 3a0a 0a60 6060 7079 7468  dies/):..```pyth
-00001850: 6f6e 0a73 7475 6479 203d 2064 6174 6162  on.study = datab
-00001860: 6173 652e 6765 745f 7374 7564 7928 2747  ase.get_study('G
-00001870: 5345 3936 3538 335f 6261 7463 6832 2729  SE96583_batch2')
-00001880: 0a73 7475 6479 0a60 6060 0a3e 6060 600a  .study.```.>```.
-00001890: 3e20 5374 7564 7928 6964 3d22 4753 4539  > Study(id="GSE9
-000018a0: 3635 3833 5f62 6174 6368 3222 2c0a 3e20  6583_batch2",.> 
-000018b0: 2020 2020 2020 7469 746c 653d 224d 756c        title="Mul
-000018c0: 7469 706c 6578 6564 2064 726f 706c 6574  tiplexed droplet
-000018d0: 2073 696e 676c 652d 6365 6c6c 2052 4e41   single-cell RNA
-000018e0: 2d73 6571 7565 6e63 696e 6720 7573 696e  -sequencing usin
-000018f0: 6720 6e61 7475 7261 6c20 6765 6e65 7469  g natural geneti
-00001900: 6320 7661 7269 6174 696f 6e20 2842 6174  c variation (Bat
-00001910: 6368 2032 2922 2c0a 3e20 2020 2020 2020  ch 2)",.>       
-00001920: 7265 6665 7265 6e63 653d 2268 7474 7073  reference="https
-00001930: 3a2f 2f77 7777 2e6e 6174 7572 652e 636f  ://www.nature.co
-00001940: 6d2f 6172 7469 636c 6573 2f6e 6274 2e34  m/articles/nbt.4
-00001950: 3034 3222 290a 0a23 2323 2054 616b 6520  042")..### Take 
-00001960: 6120 7065 656b 2061 7420 7374 7564 7920  a peek at study 
-00001970: 6d65 7461 6461 7461 3a0a 0a60 6060 7079  metadata:..```py
-00001980: 7468 6f6e 0a73 7475 6479 2e6d 6574 616c  thon.study.metal
-00001990: 6973 740a 6060 600a 3e60 6060 0a3e 205b  ist.```.>```.> [
-000019a0: 4d65 7461 6461 7461 2869 643d 302c 6e61  Metadata(id=0,na
-000019b0: 6d65 3d22 4e75 6d62 6572 206f 6620 6d52  me="Number of mR
-000019c0: 4e41 2074 7261 6e73 6372 6970 7473 222c  NA transcripts",
-000019d0: 7479 7065 3d22 4e75 6d65 7269 6322 292c  type="Numeric"),
-000019e0: 0a3e 2020 4d65 7461 6461 7461 2869 643d  .>  Metadata(id=
-000019f0: 312c 6e61 6d65 3d22 4e75 6d62 6572 206f  1,name="Number o
-00001a00: 6620 6765 6e65 7322 2c74 7970 653d 224e  f genes",type="N
-00001a10: 756d 6572 6963 2229 2c0a 3e20 204d 6574  umeric"),.>  Met
-00001a20: 6164 6174 6128 6964 3d32 2c6e 616d 653d  adata(id=2,name=
-00001a30: 2242 6174 6368 2069 6422 2c74 7970 653d  "Batch id",type=
-00001a40: 2243 6174 6567 6f72 7922 292c 0a3e 2020  "Category"),.>  
-00001a50: 4d65 7461 6461 7461 2869 643d 332c 6e61  Metadata(id=3,na
-00001a60: 6d65 3d22 5374 696d 756c 6174 696f 6e22  me="Stimulation"
-00001a70: 2c74 7970 653d 2243 6174 6567 6f72 7922  ,type="Category"
-00001a80: 292c 0a3e 2020 4d65 7461 6461 7461 2869  ),.>  Metadata(i
-00001a90: 643d 342c 6e61 6d65 3d22 4175 7468 6f72  d=4,name="Author
-00001aa0: 2773 2063 656c 6c20 7479 7065 222c 7479  's cell type",ty
-00001ab0: 7065 3d22 4361 7465 676f 7279 2229 5d0a  pe="Category")].
-00001ac0: 0a23 2323 2046 6574 6368 2061 2073 7475  .### Fetch a stu
-00001ad0: 6479 206d 6574 6164 6174 613a 0a0a 6060  dy metadata:..``
-00001ae0: 6070 7974 686f 6e0a 6d65 7461 6461 7461  `python.metadata
-00001af0: 203d 2073 7475 6479 2e6d 6574 616c 6973   = study.metalis
-00001b00: 745b 345d 0a6d 6574 6164 6174 610a 6060  t[4].metadata.``
-00001b10: 600a 3e60 6060 0a3e 4d65 7461 6461 7461  `.>```.>Metadata
-00001b20: 2869 643d 342c 6e61 6d65 3d22 4175 7468  (id=4,name="Auth
-00001b30: 6f72 2773 2063 656c 6c20 7479 7065 222c  or's cell type",
-00001b40: 7479 7065 3d22 4361 7465 676f 7279 2229  type="Category")
-00001b50: 0a60 6060 7079 7468 6f6e 0a6d 6574 6164  .```python.metad
-00001b60: 6174 612e 6665 7463 6828 290a 6d65 7461  ata.fetch().meta
-00001b70: 6461 7461 2e76 616c 7565 730a 6060 600a  data.values.```.
-00001b80: 3e60 6060 0a3e 2061 7272 6179 285b 2743  >```.> array(['C
-00001b90: 4438 2054 2063 656c 6c73 272c 2027 4465  D8 T cells', 'De
-00001ba0: 6e64 7269 7469 6320 6365 6c6c 7327 2c20  ndritic cells', 
-00001bb0: 2743 4434 2054 2063 656c 6c73 272c 202e  'CD4 T cells', .
-00001bc0: 2e2e 2c0a 3e20 2020 2020 2020 2027 4344  ..,.>        'CD
-00001bd0: 3820 5420 6365 6c6c 7327 2c20 2742 2063  8 T cells', 'B c
-00001be0: 656c 6c73 272c 2027 4344 3420 5420 6365  ells', 'CD4 T ce
-00001bf0: 6c6c 7327 5d2c 2064 7479 7065 3d27 3c55  lls'], dtype='<U
-00001c00: 3137 2729 0a0a 2323 2320 5175 6572 7920  17')..### Query 
-00001c10: 6765 6e65 733a 0a0a 6060 6070 7974 686f  genes:..```pytho
-00001c20: 6e0a 6578 705f 6d74 7820 3d20 7374 7564  n.exp_mtx = stud
-00001c30: 792e 7175 6572 795f 6765 6e65 7328 5b27  y.query_genes(['
-00001c40: 4344 3344 272c 2027 4344 3345 275d 2c20  CD3D', 'CD3E'], 
-00001c50: 6269 6f66 6c65 782e 554e 4954 5f4c 4f47  bioflex.UNIT_LOG
-00001c60: 4e4f 524d 290a 6578 705f 6d74 780a 6060  NORM).exp_mtx.``
-00001c70: 600a 3e60 6060 0a3e 203c 3239 3036 3578  `.>```.> <29065x
-00001c80: 3220 7370 6172 7365 206d 6174 7269 7820  2 sparse matrix 
-00001c90: 6f66 2074 7970 6520 273c 636c 6173 7320  of type '<class 
-00001ca0: 276e 756d 7079 2e66 6c6f 6174 3332 273e  'numpy.float32'>
-00001cb0: 270a 3e20 2020 2020 7769 7468 2031 3534  '.>     with 154
-00001cc0: 3932 2073 746f 7265 6420 656c 656d 656e  92 stored elemen
-00001cd0: 7473 2069 6e20 436f 6d70 7265 7373 6564  ts in Compressed
-00001ce0: 2053 7061 7273 6520 436f 6c75 6d6e 2066   Sparse Column f
-00001cf0: 6f72 6d61 743e 0a0a 2323 2320 4765 7420  ormat>..### Get 
-00001d00: 7374 7564 7920 6261 7263 6f64 6573 3a0a  study barcodes:.
-00001d10: 0a60 6060 7079 7468 6f6e 0a73 7475 6479  .```python.study
-00001d20: 2e62 6172 636f 6465 7328 290a 6060 600a  .barcodes().```.
-00001d30: 3e60 6060 0a3e 205b 2747 534d 3235 3630  >```.> ['GSM2560
-00001d40: 3234 395f 4141 4143 4154 4143 4341 4147  249_AAACATACCAAG
-00001d50: 4354 2d31 272c 0a3e 2020 2747 534d 3235  CT-1',.>  'GSM25
-00001d60: 3630 3234 395f 4141 4143 4154 4143 4343  60249_AAACATACCC
-00001d70: 4354 4143 2d31 272c 0a3e 2020 2e2e 2e0a  CTAC-1',.>  ....
-00001d80: 3e20 2027 4753 4d32 3536 3032 3439 5f41  >  'GSM2560249_A
-00001d90: 4154 5447 5447 4154 5443 4143 542d 3127  ATTGTGATTCACT-1'
-00001da0: 2c0a 3e20 2027 4753 4d32 3536 3032 3439  ,.>  'GSM2560249
-00001db0: 5f41 4154 5447 5447 4154 5454 4347 542d  _AATTGTGATTTCGT-
-00001dc0: 3127 2c0a 3e20 202e 2e2e 5d0a 0a23 2323  1',.>  ...]..###
-00001dd0: 2047 6574 2073 7475 6479 2066 6561 7475   Get study featu
-00001de0: 7265 733a 0a0a 6060 6070 7974 686f 6e0a  res:..```python.
-00001df0: 7374 7564 792e 6665 6174 7572 6573 2829  study.features()
-00001e00: 0a60 6060 0a3e 6060 600a 3e20 5b27 3553  .```.>```.> ['5S
-00001e10: 5f52 524e 4127 2c0a 3e20 2027 355f 3853  _RRNA',.>  '5_8S
-00001e20: 5f52 524e 4127 2c0a 3e20 202e 2e2e 0a3e  _RRNA',.>  ....>
-00001e30: 2020 2741 4330 3036 3237 3327 2c0a 3e20    'AC006273',.> 
-00001e40: 2027 4143 3030 3632 3737 272c 0a3e 2020   'AC006277',.>  
-00001e50: 2e2e 2e5d 0a0a 2323 2320 4765 7420 7374  ...]..### Get st
-00001e60: 7564 7920 6675 6c6c 206d 6174 7269 783a  udy full matrix:
-00001e70: 0a0a 6060 6070 7974 686f 6e0a 7374 7564  ..```python.stud
-00001e80: 792e 6d61 7472 6978 2862 696f 666c 6578  y.matrix(bioflex
-00001e90: 2e55 4e49 545f 4c4f 474e 4f52 4d29 0a60  .UNIT_LOGNORM).`
-00001ea0: 6060 0a3e 6060 600a 3e20 3c32 3930 3635  ``.>```.> <29065
-00001eb0: 7836 3436 3432 2073 7061 7273 6520 6d61  x64642 sparse ma
-00001ec0: 7472 6978 206f 6620 7479 7065 2027 3c63  trix of type '<c
-00001ed0: 6c61 7373 2027 6e75 6d70 792e 666c 6f61  lass 'numpy.floa
-00001ee0: 7433 3227 3e27 0a3e 2009 7769 7468 2031  t32'>'.> .with 1
-00001ef0: 3735 3730 3733 3920 7374 6f72 6564 2065  7570739 stored e
-00001f00: 6c65 6d65 6e74 7320 696e 2043 6f6d 7072  lements in Compr
-00001f10: 6573 7365 6420 5370 6172 7365 2043 6f6c  essed Sparse Col
-00001f20: 756d 6e20 666f 726d 6174 3e0a 0a23 2323  umn format>..###
-00001f30: 2045 7870 6f72 7420 5374 7564 793a 0a0a   Export Study:..
-00001f40: 6060 6070 7974 686f 6e0a 7374 7564 792e  ```python.study.
-00001f50: 6578 706f 7274 5f73 7475 6479 2862 696f  export_study(bio
-00001f60: 666c 6578 2e45 5850 4f52 545f 4835 4144  flex.EXPORT_H5AD
-00001f70: 290a 6060 600a 3e60 6060 0a3e 7b27 646f  ).```.>```.>{'do
-00001f80: 776e 6c6f 6164 5f6c 696e 6b27 3a20 2768  wnload_link': 'h
-00001f90: 7474 7073 3a2f 2f74 616c 6b32 6461 7461  ttps://talk2data
-00001fa0: 2e62 696f 7475 7269 6e67 2e63 6f6d 2f61  .bioturing.com/a
-00001fb0: 7069 2f65 7870 6f72 742f 6131 3030 3362  pi/export/a1003b
-00001fc0: 6164 3364 6431 3436 6232 3863 3762 6461  ad3dd146b28c7bda
-00001fd0: 3931 3361 3266 6333 6630 272c 0a3e 2027  913a2fc3f0',.> '
-00001fe0: 7374 7564 795f 6861 7368 5f69 6427 3a20  study_hash_id': 
-00001ff0: 2747 5345 3936 3538 335f 6261 7463 6832  'GSE96583_batch2
-00002000: 277d 0a0a 2d2d 2d2d 0a46 6f72 2066 7572  '}..----.For fur
-00002010: 7468 6572 2069 6e66 6f72 6d61 7469 6f6e  ther information
-00002020: 2070 6c65 6173 6520 6368 6563 6b20 7468   please check th
-00002030: 6520 5b64 6f63 756d 656e 7461 7469 6f6e  e [documentation
-00002040: 5d28 6874 7470 733a 2f2f 636f 6c61 622e  ](https://colab.
-00002050: 6269 6f74 7572 696e 672e 636f 6d2f 292e  bioturing.com/).
-00002060: 0a0a                                     ..
+00000290: 2320 5573 6167 6520 6269 6f63 6f6c 6162  # Usage biocolab
+000002a0: 7364 6b3a 0a0a 2323 2031 2e20 5573 6167  sdk:..## 1. Usag
+000002b0: 6520 6262 726f 7773 6572 7820 696e 2062  e bbrowserx in b
+000002c0: 696f 636f 6c61 6273 646b 3a0a 0a2a 2a54  iocolabsdk:..**T
+000002d0: 6865 2070 6163 6b61 6765 206f 6e6c 7920  he package only 
+000002e0: 616c 6c6f 7773 2064 6174 6120 7375 626d  allows data subm
+000002f0: 6973 7369 6f6e 2076 6961 2042 696f 436f  ission via BioCo
+00000300: 6c61 6220 7072 6976 6174 6520 7365 7276  lab private serv
+00000310: 6572 2e20 506c 6561 7365 2063 6f6e 6669  er. Please confi
+00000320: 6775 7265 2079 6f75 7220 746f 6b65 6e73  gure your tokens
+00000330: 2069 6e20 7468 6520 6055 7365 7220 5365   in the `User Se
+00000340: 7474 696e 6773 6020 7061 6765 2e2a 2a0a  ttings` page.**.
+00000350: 0a3c 2f62 723e 0a3c 696d 6720 616c 743d  .</br>.<img alt=
+00000360: 2273 6574 7469 6e67 2220 7372 633d 2268  "setting" src="h
+00000370: 7474 7073 3a2f 2f63 646e 2e62 696f 7475  ttps://cdn.biotu
+00000380: 7269 6e67 2e63 6f6d 2f64 6f63 756d 656e  ring.com/documen
+00000390: 7461 7469 6f6e 2f6d 642f 7573 6572 2d73  tation/md/user-s
+000003a0: 6574 7469 6e67 2e70 6e67 2220 7769 6474  etting.png" widt
+000003b0: 683d 2231 3030 2522 3e0a 3c2f 6272 3e0a  h="100%">.</br>.
+000003c0: 0a23 2323 2031 2e31 2e20 5465 7374 2074  .### 1.1. Test t
+000003d0: 6865 2063 6f6e 6e65 6374 696f 6e20 746f  he connection to
+000003e0: 2042 4272 6f77 7365 7258 2070 7269 7661   BBrowserX priva
+000003f0: 7465 2073 6572 7665 723a 0a60 6060 0a23  te server:.```.#
+00000400: 2065 7861 6d70 6c65 2e70 790a 0a66 726f   example.py..fro
+00000410: 6d20 6269 6f63 6f6c 6162 7364 6b2e 636f  m biocolabsdk.co
+00000420: 6e6e 6563 746f 7220 696d 706f 7274 2045  nnector import E
+00000430: 436f 6e6e 6563 746f 720a 0a63 6f6e 6e65  Connector..conne
+00000440: 6374 6f72 203d 2045 436f 6e6e 6563 746f  ctor = EConnecto
+00000450: 7228 0a20 2070 7269 7661 7465 5f68 6f73  r(.  private_hos
+00000460: 743d 2268 7474 7073 3a2f 2f79 6f75 7263  t="https://yourc
+00000470: 6f6d 7061 6e79 2f74 3264 5f69 6e64 6578  ompany/t2d_index
+00000480: 5f74 6f6f 6c2f 2c0a 2020 7072 6976 6174  _tool/,.  privat
+00000490: 655f 746f 6b65 6e3d 223c 696e 7075 7420  e_token="<input 
+000004a0: 796f 7572 2074 6f6b 656e 2068 6572 653e  your token here>
+000004b0: 220a 290a 0a63 6f6e 6e65 6374 6f72 2e67  ".)..connector.g
+000004c0: 6574 5f62 6272 6f77 7365 7278 2829 2e74  et_bbrowserx().t
+000004d0: 6573 745f 636f 6e6e 6563 7469 6f6e 2829  est_connection()
+000004e0: 0a60 6060 0a0a 4578 616d 706c 6520 6f75  .```..Example ou
+000004f0: 7470 7574 3a0a 0a60 6060 0a43 6f6e 6e65  tput:..```.Conne
+00000500: 6374 696e 6720 746f 2068 6f73 7420 6174  cting to host at
+00000510: 2068 7474 7073 3a2f 2f79 6f75 7263 6f6d   https://yourcom
+00000520: 7061 6e79 2f74 3264 5f69 6e64 6578 5f74  pany/t2d_index_t
+00000530: 6f6f 6c2f 6170 692f 7631 2f74 6573 745f  ool/api/v1/test_
+00000540: 636f 6e6e 6563 7469 6f6e 0a43 6f6e 6e65  connection.Conne
+00000550: 6374 696f 6e20 7375 6363 6573 7366 756c  ction successful
+00000560: 0a60 6060 0a0a 2323 2320 312e 322e 2047  .```..### 1.2. G
+00000570: 6574 2075 7365 7220 6772 6f75 7073 2061  et user groups a
+00000580: 7661 696c 6162 6c65 2066 6f72 2079 6f75  vailable for you
+00000590: 7220 746f 6b65 6e20 696e 2042 4272 6f77  r token in BBrow
+000005a0: 7365 7258 2070 7269 7661 7465 2073 6572  serX private ser
+000005b0: 7665 723a 0a60 6060 0a23 2065 7861 6d70  ver:.```.# examp
+000005c0: 6c65 2e70 790a 0a66 726f 6d20 6269 6f63  le.py..from bioc
+000005d0: 6f6c 6162 7364 6b2e 636f 6e6e 6563 746f  olabsdk.connecto
+000005e0: 7220 696d 706f 7274 2045 436f 6e6e 6563  r import EConnec
+000005f0: 746f 720a 0a63 6f6e 6e65 6374 6f72 203d  tor..connector =
+00000600: 2045 436f 6e6e 6563 746f 7228 0a20 2070   EConnector(.  p
+00000610: 7269 7661 7465 5f68 6f73 743d 2268 7474  rivate_host="htt
+00000620: 7073 3a2f 2f79 6f75 7263 6f6d 7061 6e79  ps://yourcompany
+00000630: 2f74 3264 5f69 6e64 6578 5f74 6f6f 6c2f  /t2d_index_tool/
+00000640: 2c0a 2020 7072 6976 6174 655f 746f 6b65  ,.  private_toke
+00000650: 6e3d 223c 696e 7075 7420 796f 7572 2074  n="<input your t
+00000660: 6f6b 656e 2068 6572 653e 220a 290a 0a75  oken here>".)..u
+00000670: 7365 725f 6772 6f75 7073 203d 2063 6f6e  ser_groups = con
+00000680: 6e65 6374 6f72 2e67 6574 5f62 6272 6f77  nector.get_bbrow
+00000690: 7365 7278 2829 2e67 6574 5f75 7365 725f  serx().get_user_
+000006a0: 6772 6f75 7073 2829 0a70 7269 6e74 2875  groups().print(u
+000006b0: 7365 725f 6772 6f75 7073 290a 6060 600a  ser_groups).```.
+000006c0: 0a45 7861 6d70 6c65 206f 7574 7075 743a  .Example output:
+000006d0: 0a0a 6060 600a 5b7b 2769 6427 3a20 2761  ..```.[{'id': 'a
+000006e0: 6c6c 5f6d 656d 6265 7273 272c 2027 6e61  ll_members', 'na
+000006f0: 6d65 273a 2027 416c 6c20 6d65 6d62 6572  me': 'All member
+00000700: 7327 7d2c 207b 2769 6427 3a20 2770 6572  s'}, {'id': 'per
+00000710: 736f 6e61 6c27 2c20 276e 616d 6527 3a20  sonal', 'name': 
+00000720: 2750 6572 736f 6e61 6c20 776f 726b 7370  'Personal worksp
+00000730: 6163 6527 7d5d 0a60 6060 0a0a 2323 2320  ace'}].```..### 
+00000740: 312e 332e 2053 7562 6d69 7420 796f 7572  1.3. Submit your
+00000750: 2064 6174 6120 746f 2042 4272 6f77 7365   data to BBrowse
+00000760: 7258 2070 7269 7661 7465 2073 6572 7665  rX private serve
+00000770: 723a 0a60 6060 0a23 2065 7861 6d70 6c65  r:.```.# example
+00000780: 2e70 790a 6672 6f6d 2062 696f 636f 6c61  .py.from biocola
+00000790: 6273 646b 2e63 6f6e 6e65 6374 6f72 2069  bsdk.connector i
+000007a0: 6d70 6f72 7420 4543 6f6e 6e65 6374 6f72  mport EConnector
+000007b0: 0a66 726f 6d20 6269 6f74 7572 696e 675f  .from bioturing_
+000007c0: 636f 6e6e 6563 746f 722e 7479 7069 6e67  connector.typing
+000007d0: 2069 6d70 6f72 7420 496e 7075 744d 6174   import InputMat
+000007e0: 7269 7854 7970 650a 6672 6f6d 2062 696f  rixType.from bio
+000007f0: 7475 7269 6e67 5f63 6f6e 6e65 6374 6f72  turing_connector
+00000800: 2e74 7970 696e 6720 696d 706f 7274 2053  .typing import S
+00000810: 7065 6369 6573 0a66 726f 6d20 6269 6f74  pecies.from biot
+00000820: 7572 696e 675f 636f 6e6e 6563 746f 722e  uring_connector.
+00000830: 7479 7069 6e67 2069 6d70 6f72 7420 5374  typing import St
+00000840: 7564 7954 7970 650a 0a63 6f6e 6e65 6374  udyType..connect
+00000850: 6f72 203d 2045 436f 6e6e 6563 746f 7228  or = EConnector(
+00000860: 0a20 2070 7269 7661 7465 5f68 6f73 743d  .  private_host=
+00000870: 2268 7474 7073 3a2f 2f79 6f75 7263 6f6d  "https://yourcom
+00000880: 7061 6e79 2f74 3264 5f69 6e64 6578 5f74  pany/t2d_index_t
+00000890: 6f6f 6c2f 2c0a 2020 7072 6976 6174 655f  ool/,.  private_
+000008a0: 746f 6b65 6e3d 223c 696e 7075 7420 796f  token="<input yo
+000008b0: 7572 2074 6f6b 656e 2068 6572 653e 220a  ur token here>".
+000008c0: 290a 0a23 2043 616c 6c20 7468 6973 2066  )..# Call this f
+000008d0: 756e 6374 696f 6e20 6669 7273 7420 746f  unction first to
+000008e0: 2067 6574 2061 7661 696c 6162 6c65 2067   get available g
+000008f0: 726f 7570 7320 616e 6420 7468 6569 7220  roups and their 
+00000900: 6964 2e0a 7573 6572 5f67 726f 7570 7320  id..user_groups 
+00000910: 3d20 636f 6e6e 6563 746f 722e 6765 745f  = connector.get_
+00000920: 6262 726f 7773 6572 7828 292e 6765 745f  bbrowserx().get_
+00000930: 7573 6572 5f67 726f 7570 7328 290a 2320  user_groups().# 
+00000940: 4578 616d 706c 653a 2075 7365 725f 6772  Example: user_gr
+00000950: 6f75 7073 2069 7320 6e6f 7720 5b7b 2769  oups is now [{'i
+00000960: 6427 3a20 2761 6c6c 5f6d 656d 6265 7273  d': 'all_members
+00000970: 272c 2027 6e61 6d65 273a 2027 416c 6c20  ', 'name': 'All 
+00000980: 6d65 6d62 6572 7327 7d2c 207b 2769 6427  members'}, {'id'
+00000990: 3a20 2770 6572 736f 6e61 6c27 2c20 276e  : 'personal', 'n
+000009a0: 616d 6527 3a20 2750 6572 736f 6e61 6c20  ame': 'Personal 
+000009b0: 776f 726b 7370 6163 6527 7d5d 0a0a 2320  workspace'}]..# 
+000009c0: 5375 626d 6974 7469 6e67 2074 6865 2073  Submitting the s
+000009d0: 6361 6e70 7920 6f62 6a65 6374 2066 726f  canpy object fro
+000009e0: 6d20 7333 3a0a 6261 7463 685f 696e 666f  m s3:.batch_info
+000009f0: 203d 205b 7b0a 2020 2020 276d 6174 7269   = [{.    'matri
+00000a00: 7827 3a20 2773 335f 7061 7468 2f47 5345  x': 's3_path/GSE
+00000a10: 3132 3832 3233 5f31 2e68 3561 6427 2c0a  128223_1.h5ad',.
+00000a20: 7d2c 207b 0a20 2020 2027 6d61 7472 6978  }, {.    'matrix
+00000a30: 273a 2027 7333 5f70 6174 682f 4753 4531  ': 's3_path/GSE1
+00000a40: 3238 3232 335f 322e 6835 6164 272c 0a7d  28223_2.h5ad',.}
+00000a50: 5d0a 0a63 6f6e 6e65 6374 6f72 2e67 6574  ]..connector.get
+00000a60: 5f62 6272 6f77 7365 7278 2829 2e73 7562  _bbrowserx().sub
+00000a70: 6d69 745f 7374 7564 795f 6672 6f6d 5f73  mit_study_from_s
+00000a80: 3328 0a20 2067 726f 7570 5f69 643d 2770  3(.  group_id='p
+00000a90: 6572 736f 6e61 6c27 2c0a 2020 6261 7463  ersonal',.  batc
+00000aa0: 685f 696e 666f 3d62 6174 6368 5f69 6e66  h_info=batch_inf
+00000ab0: 6f2c 0a20 2073 7475 6479 5f69 643d 2747  o,.  study_id='G
+00000ac0: 5345 3132 3832 3233 272c 0a20 206e 616d  SE128223',.  nam
+00000ad0: 653d 2754 6869 7320 6973 206d 7920 6669  e='This is my fi
+00000ae0: 7273 7420 7374 7564 7927 2c0a 2020 6175  rst study',.  au
+00000af0: 7468 6f72 733d 5b27 4875 7920 4e67 7579  thors=['Huy Nguy
+00000b00: 656e 275d 2c0a 2020 7370 6563 6965 733d  en'],.  species=
+00000b10: 5370 6563 6965 732e 4855 4d41 4e2e 7661  Species.HUMAN.va
+00000b20: 6c75 652c 0a20 2069 6e70 7574 5f6d 6174  lue,.  input_mat
+00000b30: 7269 785f 7479 7065 3d49 6e70 7574 4d61  rix_type=InputMa
+00000b40: 7472 6978 5479 7065 2e52 4157 2e76 616c  trixType.RAW.val
+00000b50: 7565 2c0a 2020 7374 7564 795f 7479 7065  ue,.  study_type
+00000b60: 3d53 7475 6479 5479 7065 2e48 3541 442e  =StudyType.H5AD.
+00000b70: 7661 6c75 650a 290a 0a23 2053 7562 6d69  value.)..# Submi
+00000b80: 7474 696e 6720 7468 6520 7363 616e 7079  tting the scanpy
+00000b90: 206f 626a 6563 7420 6672 6f6d 206c 6f63   object from loc
+00000ba0: 616c 206d 6163 6869 6e65 3a0a 6261 7463  al machine:.batc
+00000bb0: 685f 696e 666f 203d 205b 7b0a 2020 2020  h_info = [{.    
+00000bc0: 276d 6174 7269 7827 3a20 276c 6f63 616c  'matrix': 'local
+00000bd0: 5f70 6174 682f 4753 4531 3238 3232 335f  _path/GSE128223_
+00000be0: 312e 6835 6164 272c 0a7d 2c20 7b0a 2020  1.h5ad',.}, {.  
+00000bf0: 2020 276d 6174 7269 7827 3a20 276c 6f63    'matrix': 'loc
+00000c00: 616c 5f70 6174 682f 4753 4531 3238 3232  al_path/GSE12822
+00000c10: 335f 322e 6835 6164 272c 0a7d 5d0a 0a63  3_2.h5ad',.}]..c
+00000c20: 6f6e 6e65 6374 6f72 2e67 6574 5f62 6272  onnector.get_bbr
+00000c30: 6f77 7365 7278 2829 2e73 7562 6d69 745f  owserx().submit_
+00000c40: 7374 7564 795f 6672 6f6d 5f6c 6f63 616c  study_from_local
+00000c50: 280a 2020 6772 6f75 705f 6964 3d27 7065  (.  group_id='pe
+00000c60: 7273 6f6e 616c 272c 0a20 2062 6174 6368  rsonal',.  batch
+00000c70: 5f69 6e66 6f3d 6261 7463 685f 696e 666f  _info=batch_info
+00000c80: 2c0a 2020 7374 7564 795f 6964 3d27 4753  ,.  study_id='GS
+00000c90: 4531 3238 3232 3327 2c0a 2020 6e61 6d65  E128223',.  name
+00000ca0: 3d27 5468 6973 2069 7320 6d79 2066 6972  ='This is my fir
+00000cb0: 7374 2073 7475 6479 272c 0a20 2061 7574  st study',.  aut
+00000cc0: 686f 7273 3d5b 2748 7579 204e 6775 7965  hors=['Huy Nguye
+00000cd0: 6e27 5d2c 0a20 2073 7065 6369 6573 3d53  n'],.  species=S
+00000ce0: 7065 6369 6573 2e48 554d 414e 2e76 616c  pecies.HUMAN.val
+00000cf0: 7565 2c0a 2020 696e 7075 745f 6d61 7472  ue,.  input_matr
+00000d00: 6978 5f74 7970 653d 496e 7075 744d 6174  ix_type=InputMat
+00000d10: 7269 7854 7970 652e 5241 572e 7661 6c75  rixType.RAW.valu
+00000d20: 652c 0a20 2073 7475 6479 5f74 7970 653d  e,.  study_type=
+00000d30: 5374 7564 7954 7970 652e 4835 4144 2e76  StudyType.H5AD.v
+00000d40: 616c 7565 0a29 0a0a 2323 2032 2e20 5573  alue.)..## 2. Us
+00000d50: 6167 6520 7461 6c6b 3264 6174 6120 696e  age talk2data in
+00000d60: 2062 696f 636f 6c61 6273 646b 3a0a 0a60   biocolabsdk:..`
+00000d70: 6060 0a23 2065 7861 6d70 6c65 2e70 790a  ``.# example.py.
+00000d80: 0a66 726f 6d20 6269 6f63 6f6c 6162 7364  .from biocolabsd
+00000d90: 6b2e 636f 6e6e 6563 746f 7220 696d 706f  k.connector impo
+00000da0: 7274 2045 436f 6e6e 6563 746f 720a 0a63  rt EConnector..c
+00000db0: 6f6e 6e65 6374 6f72 203d 2045 436f 6e6e  onnector = EConn
+00000dc0: 6563 746f 7228 0a20 2070 7562 6c69 635f  ector(.  public_
+00000dd0: 746f 6b65 6e3d 223c 696e 7075 7420 796f  token="<input yo
+00000de0: 7572 2074 6f6b 656e 2068 6572 653e 220a  ur token here>".
+00000df0: 290a 0a63 6f6e 6e65 6374 6f72 2e67 6574  )..connector.get
+00000e00: 5f74 616c 6b32 6461 7461 2829 2e74 6573  _talk2data().tes
+00000e10: 745f 636f 6e6e 6563 7469 6f6e 2829 0a60  t_connection().`
+00000e20: 6060 0a0a 4578 616d 706c 6520 6f75 7470  ``..Example outp
+00000e30: 7574 3a0a 0a60 6060 0a43 6f6e 6e65 6374  ut:..```.Connect
+00000e40: 696e 6720 746f 2068 6f73 7420 6174 2068  ing to host at h
+00000e50: 7474 7073 3a2f 2f74 616c 6b32 6461 7461  ttps://talk2data
+00000e60: 2e62 696f 7475 7269 6e67 2e63 6f6d 2f74  .bioturing.com/t
+00000e70: 3264 5f69 6e64 6578 5f74 6f6f 6c2f 6170  2d_index_tool/ap
+00000e80: 692f 7631 2f74 6573 745f 636f 6e6e 6563  i/v1/test_connec
+00000e90: 7469 6f6e 0a43 6f6e 6e65 6374 696f 6e20  tion.Connection 
+00000ea0: 7375 6363 6573 7366 756c 0a60 6060 0a0a  successful.```..
+00000eb0: 596f 7520 6361 6e20 7574 696c 697a 6520  You can utilize 
+00000ec0: 616c 6c20 7468 6520 6665 6174 7572 6573  all the features
+00000ed0: 2069 6e20 6120 7369 6d69 6c61 7220 7761   in a similar wa
+00000ee0: 7920 746f 2061 6363 6573 7369 6e67 2061  y to accessing a
+00000ef0: 2042 4272 6f73 6572 5820 7072 6976 6174   BBroserX privat
+00000f00: 6520 7365 7276 6572 2e0a 0a23 2320 332e  e server...## 3.
+00000f10: 2055 7361 6765 2062 696f 666c 6578 2069   Usage bioflex i
+00000f20: 6e20 6269 6f63 6f6c 6162 7364 6b3a 0a0a  n biocolabsdk:..
+00000f30: 596f 7520 6361 6e20 6765 7420 6120 4269  You can get a Bi
+00000f40: 6f66 6c65 7820 746f 6b65 6e20 6174 3a20  oflex token at: 
+00000f50: 0a3c 6120 6872 6566 3d22 6874 7470 733a  .<a href="https:
+00000f60: 2f2f 636f 6c61 622e 6269 6f74 7572 696e  //colab.bioturin
+00000f70: 672e 636f 6d2f 7265 7175 6573 742f 6269  g.com/request/bi
+00000f80: 6f66 6c65 782f 746f 6b65 6e22 3e68 7474  oflex/token">htt
+00000f90: 7073 3a2f 2f63 6f6c 6162 2e62 696f 7475  ps://colab.biotu
+00000fa0: 7269 6e67 2e63 6f6d 2f72 6571 7565 7374  ring.com/request
+00000fb0: 2f62 696f 666c 6578 2f74 6f6b 656e 3c2f  /bioflex/token</
+00000fc0: 613e 0a0a 2323 2320 4372 6561 7465 2061  a>..### Create a
+00000fd0: 2063 6f6e 6e65 6374 696f 6e20 7573 696e   connection usin
+00000fe0: 6720 6163 6365 7373 2074 6f6b 656e 3a0a  g access token:.
+00000ff0: 0a60 6060 7079 7468 6f6e 0a66 726f 6d20  .```python.from 
+00001000: 6269 6f63 6f6c 6162 7364 6b2e 636f 6e6e  biocolabsdk.conn
+00001010: 6563 746f 7220 696d 706f 7274 2045 436f  ector import ECo
+00001020: 6e6e 6563 746f 720a 0a63 6f6e 6e65 6374  nnector..connect
+00001030: 6f72 203d 2045 436f 6e6e 6563 746f 7228  or = EConnector(
+00001040: 0a20 2062 696f 666c 6578 5f74 6f6b 656e  .  bioflex_token
+00001050: 3d22 3c69 6e70 7574 2079 6f75 7220 746f  ="<input your to
+00001060: 6b65 6e20 6865 7265 3e22 0a29 0a60 6060  ken here>".).```
+00001070: 0a0a 2323 2320 4c69 7374 2061 7661 696c  ..### List avail
+00001080: 6162 6c65 2064 6174 6162 6173 6573 3a0a  able databases:.
+00001090: 0a60 6060 7079 7468 6f6e 0a64 6174 6162  .```python.datab
+000010a0: 6173 6573 203d 2063 6f6e 6e65 6374 6f72  ases = connector
+000010b0: 2e67 6574 5f62 696f 666c 6578 2829 2e64  .get_bioflex().d
+000010c0: 6174 6162 6173 6573 2829 0a60 6060 0a3e  atabases().```.>
+000010d0: 6060 600a 3e20 5b44 6174 6142 6173 6528  ```.> [DataBase(
+000010e0: 6964 3d22 3530 3130 6337 6435 3733 6165  id="5010c7d573ae
+000010f0: 3466 6632 6239 3639 3134 3232 6239 3961  4ff2b9691422b99a
+00001100: 6132 6364 222c 0a3e 2020 2020 2020 2020  a2cd",.>        
+00001110: 2020 206e 616d 653d 2242 696f 5475 7269     name="BioTuri
+00001120: 6e67 2064 6174 6162 6173 6522 2c73 7065  ng database",spe
+00001130: 6369 6573 3d22 6875 6d61 6e22 2c76 6572  cies="human",ver
+00001140: 7369 6f6e 3d31 292c 0a3e 2044 6174 6142  sion=1),.> DataB
+00001150: 6173 6528 6964 3d22 3530 3130 6337 6435  ase(id="5010c7d5
+00001160: 3733 6165 3466 6632 6239 3639 3134 3232  73ae4ff2b9691422
+00001170: 6239 3961 6132 6364 222c 0a3e 2020 2020  b99aa2cd",.>    
+00001180: 2020 2020 2020 206e 616d 653d 2242 696f         name="Bio
+00001190: 5475 7269 6e67 2064 6174 6162 6173 6522  Turing database"
+000011a0: 2c73 7065 6369 6573 3d22 6875 6d61 6e22  ,species="human"
+000011b0: 2c76 6572 7369 6f6e 3d32 292c 0a3e 2044  ,version=2),.> D
+000011c0: 6174 6142 6173 6528 6964 3d22 3530 3130  ataBase(id="5010
+000011d0: 6337 6435 3733 6165 3466 6632 6239 3639  c7d573ae4ff2b969
+000011e0: 3134 3232 6239 3961 6132 6364 222c 0a3e  1422b99aa2cd",.>
+000011f0: 2020 2020 2020 2020 2020 206e 616d 653d             name=
+00001200: 2242 696f 5475 7269 6e67 2064 6174 6162  "BioTuring datab
+00001210: 6173 6522 2c73 7065 6369 6573 3d22 6875  ase",species="hu
+00001220: 6d61 6e22 2c76 6572 7369 6f6e 3d33 295d  man",version=3)]
+00001230: 0a0a 2323 2320 4765 7420 6461 7461 6261  ..### Get databa
+00001240: 7365 2063 656c 6c20 7479 7065 7320 6765  se cell types ge
+00001250: 6e65 2065 7870 7265 7373 696f 6e20 7375  ne expression su
+00001260: 6d6d 6172 790a 0a60 6060 7079 7468 6f6e  mmary..```python
+00001270: 0a64 6174 6162 6173 6520 3d20 6461 7461  .database = data
+00001280: 6261 7365 735b 325d 0a64 6174 6162 6173  bases[2].databas
+00001290: 652e 6765 745f 6365 6c6c 7479 7065 735f  e.get_celltypes_
+000012a0: 6578 7072 6573 7369 6f6e 5f73 756d 6d61  expression_summa
+000012b0: 7279 285b 2743 4433 4427 2c20 2743 4433  ry(['CD3D', 'CD3
+000012c0: 4527 5d29 0a60 6060 0a3e 6060 600a 3e20  E']).```.>```.> 
+000012d0: 7b27 4344 3344 273a 205b 5375 6d6d 6172  {'CD3D': [Summar
+000012e0: 7928 6e61 6d65 3d22 4220 6365 6c6c 222c  y(name="B cell",
+000012f0: 7375 6d3d 3730 3731 3038 3837 342e 302c  sum=707108874.0,
+00001300: 6d65 616e 3d34 3139 322e 3730 3936 2c72  mean=4192.7096,r
+00001310: 6174 653d 302e 3033 352c 636f 756e 743d  ate=0.035,count=
+00001320: 3136 3836 3532 2e30 2c74 6f74 616c 3d34  168652.0,total=4
+00001330: 3831 3239 3637 292c 0a3e 2009 5375 6d6d  812967),.> .Summ
+00001340: 6172 7928 6e61 6d65 3d22 4344 342d 706f  ary(name="CD4-po
+00001350: 7369 7469 7665 2c20 616c 7068 612d 6265  sitive, alpha-be
+00001360: 7461 2054 2063 656c 6c22 2c73 756d 3d39  ta T cell",sum=9
+00001370: 3438 3939 3837 3434 322e 302c 6d65 616e  489987442.0,mean
+00001380: 3d34 3635 372e 3536 3139 2c72 6174 653d  =4657.5619,rate=
+00001390: 302e 3532 3833 2c63 6f75 6e74 3d32 3033  0.5283,count=203
+000013a0: 3735 3434 2e30 2c74 6f74 616c 3d33 3835  7544.0,total=385
+000013b0: 3635 3930 292c 0a3e 2009 2e2e 2e0a 3e20  6590),.> .....> 
+000013c0: 0953 756d 6d61 7279 286e 616d 653d 2263  .Summary(name="c
+000013d0: 6f72 6e65 616c 2070 726f 6765 6e69 746f  orneal progenito
+000013e0: 7222 2c73 756d 3d30 2e30 2c6d 6561 6e3d  r",sum=0.0,mean=
+000013f0: 302e 302c 7261 7465 3d30 2e30 2c63 6f75  0.0,rate=0.0,cou
+00001400: 6e74 3d30 2e30 2c74 6f74 616c 3d33 3937  nt=0.0,total=397
+00001410: 3329 2c0a 3e20 0953 756d 6d61 7279 286e  3),.> .Summary(n
+00001420: 616d 653d 226e 7563 6c65 7573 2070 756c  ame="nucleus pul
+00001430: 706f 7375 7320 7072 6f67 656e 6974 6f72  posus progenitor
+00001440: 2063 656c 6c22 2c73 756d 3d30 2e30 2c6d   cell",sum=0.0,m
+00001450: 6561 6e3d 302e 302c 7261 7465 3d30 2e30  ean=0.0,rate=0.0
+00001460: 2c63 6f75 6e74 3d30 2e30 2c74 6f74 616c  ,count=0.0,total
+00001470: 3d32 3331 3029 5d7d 0a0a 0a23 2323 2043  =2310)]}...### C
+00001480: 7265 6174 6520 7374 7564 7920 696e 7374  reate study inst
+00001490: 616e 6365 2c20 7573 696e 6720 7374 7564  ance, using stud
+000014a0: 7920 6861 7368 2049 4420 6672 6f6d 205b  y hash ID from [
+000014b0: 4269 6f54 7572 696e 6720 7374 7564 6965  BioTuring studie
+000014c0: 735d 2868 7474 7073 3a2f 2f74 616c 6b32  s](https://talk2
+000014d0: 6461 7461 2e62 696f 7475 7269 6e67 2e63  data.bioturing.c
+000014e0: 6f6d 2f73 7475 6469 6573 2f29 3a0a 0a60  om/studies/):..`
+000014f0: 6060 7079 7468 6f6e 0a73 7475 6479 203d  ``python.study =
+00001500: 2064 6174 6162 6173 652e 6765 745f 7374   database.get_st
+00001510: 7564 7928 2747 5345 3936 3538 335f 6261  udy('GSE96583_ba
+00001520: 7463 6832 2729 0a73 7475 6479 0a60 6060  tch2').study.```
+00001530: 0a3e 6060 600a 3e20 5374 7564 7928 6964  .>```.> Study(id
+00001540: 3d22 4753 4539 3635 3833 5f62 6174 6368  ="GSE96583_batch
+00001550: 3222 2c0a 3e20 2020 2020 2020 7469 746c  2",.>       titl
+00001560: 653d 224d 756c 7469 706c 6578 6564 2064  e="Multiplexed d
+00001570: 726f 706c 6574 2073 696e 676c 652d 6365  roplet single-ce
+00001580: 6c6c 2052 4e41 2d73 6571 7565 6e63 696e  ll RNA-sequencin
+00001590: 6720 7573 696e 6720 6e61 7475 7261 6c20  g using natural 
+000015a0: 6765 6e65 7469 6320 7661 7269 6174 696f  genetic variatio
+000015b0: 6e20 2842 6174 6368 2032 2922 2c0a 3e20  n (Batch 2)",.> 
+000015c0: 2020 2020 2020 7265 6665 7265 6e63 653d        reference=
+000015d0: 2268 7474 7073 3a2f 2f77 7777 2e6e 6174  "https://www.nat
+000015e0: 7572 652e 636f 6d2f 6172 7469 636c 6573  ure.com/articles
+000015f0: 2f6e 6274 2e34 3034 3222 290a 0a23 2323  /nbt.4042")..###
+00001600: 2054 616b 6520 6120 7065 656b 2061 7420   Take a peek at 
+00001610: 7374 7564 7920 6d65 7461 6461 7461 3a0a  study metadata:.
+00001620: 0a60 6060 7079 7468 6f6e 0a73 7475 6479  .```python.study
+00001630: 2e6d 6574 616c 6973 740a 6060 600a 3e60  .metalist.```.>`
+00001640: 6060 0a3e 205b 4d65 7461 6461 7461 2869  ``.> [Metadata(i
+00001650: 643d 302c 6e61 6d65 3d22 4e75 6d62 6572  d=0,name="Number
+00001660: 206f 6620 6d52 4e41 2074 7261 6e73 6372   of mRNA transcr
+00001670: 6970 7473 222c 7479 7065 3d22 4e75 6d65  ipts",type="Nume
+00001680: 7269 6322 292c 0a3e 2020 4d65 7461 6461  ric"),.>  Metada
+00001690: 7461 2869 643d 312c 6e61 6d65 3d22 4e75  ta(id=1,name="Nu
+000016a0: 6d62 6572 206f 6620 6765 6e65 7322 2c74  mber of genes",t
+000016b0: 7970 653d 224e 756d 6572 6963 2229 2c0a  ype="Numeric"),.
+000016c0: 3e20 204d 6574 6164 6174 6128 6964 3d32  >  Metadata(id=2
+000016d0: 2c6e 616d 653d 2242 6174 6368 2069 6422  ,name="Batch id"
+000016e0: 2c74 7970 653d 2243 6174 6567 6f72 7922  ,type="Category"
+000016f0: 292c 0a3e 2020 4d65 7461 6461 7461 2869  ),.>  Metadata(i
+00001700: 643d 332c 6e61 6d65 3d22 5374 696d 756c  d=3,name="Stimul
+00001710: 6174 696f 6e22 2c74 7970 653d 2243 6174  ation",type="Cat
+00001720: 6567 6f72 7922 292c 0a3e 2020 4d65 7461  egory"),.>  Meta
+00001730: 6461 7461 2869 643d 342c 6e61 6d65 3d22  data(id=4,name="
+00001740: 4175 7468 6f72 2773 2063 656c 6c20 7479  Author's cell ty
+00001750: 7065 222c 7479 7065 3d22 4361 7465 676f  pe",type="Catego
+00001760: 7279 2229 5d0a 0a23 2323 2046 6574 6368  ry")]..### Fetch
+00001770: 2061 2073 7475 6479 206d 6574 6164 6174   a study metadat
+00001780: 613a 0a0a 6060 6070 7974 686f 6e0a 6d65  a:..```python.me
+00001790: 7461 6461 7461 203d 2073 7475 6479 2e6d  tadata = study.m
+000017a0: 6574 616c 6973 745b 345d 0a6d 6574 6164  etalist[4].metad
+000017b0: 6174 610a 6060 600a 3e60 6060 0a3e 4d65  ata.```.>```.>Me
+000017c0: 7461 6461 7461 2869 643d 342c 6e61 6d65  tadata(id=4,name
+000017d0: 3d22 4175 7468 6f72 2773 2063 656c 6c20  ="Author's cell 
+000017e0: 7479 7065 222c 7479 7065 3d22 4361 7465  type",type="Cate
+000017f0: 676f 7279 2229 0a60 6060 7079 7468 6f6e  gory").```python
+00001800: 0a6d 6574 6164 6174 612e 6665 7463 6828  .metadata.fetch(
+00001810: 290a 6d65 7461 6461 7461 2e76 616c 7565  ).metadata.value
+00001820: 730a 6060 600a 3e60 6060 0a3e 2061 7272  s.```.>```.> arr
+00001830: 6179 285b 2743 4438 2054 2063 656c 6c73  ay(['CD8 T cells
+00001840: 272c 2027 4465 6e64 7269 7469 6320 6365  ', 'Dendritic ce
+00001850: 6c6c 7327 2c20 2743 4434 2054 2063 656c  lls', 'CD4 T cel
+00001860: 6c73 272c 202e 2e2e 2c0a 3e20 2020 2020  ls', ...,.>     
+00001870: 2020 2027 4344 3820 5420 6365 6c6c 7327     'CD8 T cells'
+00001880: 2c20 2742 2063 656c 6c73 272c 2027 4344  , 'B cells', 'CD
+00001890: 3420 5420 6365 6c6c 7327 5d2c 2064 7479  4 T cells'], dty
+000018a0: 7065 3d27 3c55 3137 2729 0a0a 2323 2320  pe='<U17')..### 
+000018b0: 5175 6572 7920 6765 6e65 733a 0a0a 6060  Query genes:..``
+000018c0: 6070 7974 686f 6e0a 6578 705f 6d74 7820  `python.exp_mtx 
+000018d0: 3d20 7374 7564 792e 7175 6572 795f 6765  = study.query_ge
+000018e0: 6e65 7328 5b27 4344 3344 272c 2027 4344  nes(['CD3D', 'CD
+000018f0: 3345 275d 2c20 6269 6f66 6c65 782e 554e  3E'], bioflex.UN
+00001900: 4954 5f4c 4f47 4e4f 524d 290a 6578 705f  IT_LOGNORM).exp_
+00001910: 6d74 780a 6060 600a 3e60 6060 0a3e 203c  mtx.```.>```.> <
+00001920: 3239 3036 3578 3220 7370 6172 7365 206d  29065x2 sparse m
+00001930: 6174 7269 7820 6f66 2074 7970 6520 273c  atrix of type '<
+00001940: 636c 6173 7320 276e 756d 7079 2e66 6c6f  class 'numpy.flo
+00001950: 6174 3332 273e 270a 3e20 2020 2020 7769  at32'>'.>     wi
+00001960: 7468 2031 3534 3932 2073 746f 7265 6420  th 15492 stored 
+00001970: 656c 656d 656e 7473 2069 6e20 436f 6d70  elements in Comp
+00001980: 7265 7373 6564 2053 7061 7273 6520 436f  ressed Sparse Co
+00001990: 6c75 6d6e 2066 6f72 6d61 743e 0a0a 2323  lumn format>..##
+000019a0: 2320 4765 7420 7374 7564 7920 6261 7263  # Get study barc
+000019b0: 6f64 6573 3a0a 0a60 6060 7079 7468 6f6e  odes:..```python
+000019c0: 0a73 7475 6479 2e62 6172 636f 6465 7328  .study.barcodes(
+000019d0: 290a 6060 600a 3e60 6060 0a3e 205b 2747  ).```.>```.> ['G
+000019e0: 534d 3235 3630 3234 395f 4141 4143 4154  SM2560249_AAACAT
+000019f0: 4143 4341 4147 4354 2d31 272c 0a3e 2020  ACCAAGCT-1',.>  
+00001a00: 2747 534d 3235 3630 3234 395f 4141 4143  'GSM2560249_AAAC
+00001a10: 4154 4143 4343 4354 4143 2d31 272c 0a3e  ATACCCCTAC-1',.>
+00001a20: 2020 2e2e 2e0a 3e20 2027 4753 4d32 3536    ....>  'GSM256
+00001a30: 3032 3439 5f41 4154 5447 5447 4154 5443  0249_AATTGTGATTC
+00001a40: 4143 542d 3127 2c0a 3e20 2027 4753 4d32  ACT-1',.>  'GSM2
+00001a50: 3536 3032 3439 5f41 4154 5447 5447 4154  560249_AATTGTGAT
+00001a60: 5454 4347 542d 3127 2c0a 3e20 202e 2e2e  TTCGT-1',.>  ...
+00001a70: 5d0a 0a23 2323 2047 6574 2073 7475 6479  ]..### Get study
+00001a80: 2066 6561 7475 7265 733a 0a0a 6060 6070   features:..```p
+00001a90: 7974 686f 6e0a 7374 7564 792e 6665 6174  ython.study.feat
+00001aa0: 7572 6573 2829 0a60 6060 0a3e 6060 600a  ures().```.>```.
+00001ab0: 3e20 5b27 3553 5f52 524e 4127 2c0a 3e20  > ['5S_RRNA',.> 
+00001ac0: 2027 355f 3853 5f52 524e 4127 2c0a 3e20   '5_8S_RRNA',.> 
+00001ad0: 202e 2e2e 0a3e 2020 2741 4330 3036 3237   ....>  'AC00627
+00001ae0: 3327 2c0a 3e20 2027 4143 3030 3632 3737  3',.>  'AC006277
+00001af0: 272c 0a3e 2020 2e2e 2e5d 0a0a 2323 2320  ',.>  ...]..### 
+00001b00: 4765 7420 7374 7564 7920 6675 6c6c 206d  Get study full m
+00001b10: 6174 7269 783a 0a0a 6060 6070 7974 686f  atrix:..```pytho
+00001b20: 6e0a 7374 7564 792e 6d61 7472 6978 2862  n.study.matrix(b
+00001b30: 696f 666c 6578 2e55 4e49 545f 4c4f 474e  ioflex.UNIT_LOGN
+00001b40: 4f52 4d29 0a60 6060 0a3e 6060 600a 3e20  ORM).```.>```.> 
+00001b50: 3c32 3930 3635 7836 3436 3432 2073 7061  <29065x64642 spa
+00001b60: 7273 6520 6d61 7472 6978 206f 6620 7479  rse matrix of ty
+00001b70: 7065 2027 3c63 6c61 7373 2027 6e75 6d70  pe '<class 'nump
+00001b80: 792e 666c 6f61 7433 3227 3e27 0a3e 2009  y.float32'>'.> .
+00001b90: 7769 7468 2031 3735 3730 3733 3920 7374  with 17570739 st
+00001ba0: 6f72 6564 2065 6c65 6d65 6e74 7320 696e  ored elements in
+00001bb0: 2043 6f6d 7072 6573 7365 6420 5370 6172   Compressed Spar
+00001bc0: 7365 2043 6f6c 756d 6e20 666f 726d 6174  se Column format
+00001bd0: 3e0a 0a23 2323 2045 7870 6f72 7420 5374  >..### Export St
+00001be0: 7564 793a 0a0a 6060 6070 7974 686f 6e0a  udy:..```python.
+00001bf0: 7374 7564 792e 6578 706f 7274 5f73 7475  study.export_stu
+00001c00: 6479 2862 696f 666c 6578 2e45 5850 4f52  dy(bioflex.EXPOR
+00001c10: 545f 4835 4144 290a 6060 600a 3e60 6060  T_H5AD).```.>```
+00001c20: 0a3e 7b27 646f 776e 6c6f 6164 5f6c 696e  .>{'download_lin
+00001c30: 6b27 3a20 2768 7474 7073 3a2f 2f74 616c  k': 'https://tal
+00001c40: 6b32 6461 7461 2e62 696f 7475 7269 6e67  k2data.bioturing
+00001c50: 2e63 6f6d 2f61 7069 2f65 7870 6f72 742f  .com/api/export/
+00001c60: 6131 3030 3362 6164 3364 6431 3436 6232  a1003bad3dd146b2
+00001c70: 3863 3762 6461 3931 3361 3266 6333 6630  8c7bda913a2fc3f0
+00001c80: 272c 0a3e 2027 7374 7564 795f 6861 7368  ',.> 'study_hash
+00001c90: 5f69 6427 3a20 2747 5345 3936 3538 335f  _id': 'GSE96583_
+00001ca0: 6261 7463 6832 277d 0a0a 2d2d 2d2d 0a46  batch2'}..----.F
+00001cb0: 6f72 2066 7572 7468 6572 2069 6e66 6f72  or further infor
+00001cc0: 6d61 7469 6f6e 2070 6c65 6173 6520 6368  mation please ch
+00001cd0: 6563 6b20 7468 6520 5b64 6f63 756d 656e  eck the [documen
+00001ce0: 7461 7469 6f6e 5d28 6874 7470 733a 2f2f  tation](https://
+00001cf0: 636f 6c61 622e 6269 6f74 7572 696e 672e  colab.bioturing.
+00001d00: 636f 6d2f 292e 0a                        com/)..
```

